<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Expands the set of integration methods available to Seurat • SeuratIntegrate</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" sizes="any" href="favicon.ico">
<link rel="manifest" href="site.webmanifest">
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="Expands the set of integration methods available to Seurat">
<meta name="description" content="R package gathering a set of wrappers to apply various integration methods to Seurat objects (and rate such methods). Intended to apply to Seurat V5 objects bearing multiple layers.">
<meta property="og:description" content="R package gathering a set of wrappers to apply various integration methods to Seurat objects (and rate such methods). Intended to apply to Seurat V5 objects bearing multiple layers.">
<meta property="og:image" content="https://fspecque.github.io/test/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">SeuratIntegrate</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/SeuratIntegrate.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/introduction.html">Integration example: pbmcsca data</a></li>
    <li><a class="dropdown-item" href="articles/setup_and_tips.html">Advice about conda and future</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Aide-mémoires"><span class="fa fa-sticky-note"></span></button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-">
<li><a class="dropdown-item" href="articles/memo_integration.html">Integration methods</a></li>
    <li><a class="dropdown-item" href="articles/memo_score.html">Scoring metrics</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cbib/Seurat-Integrate/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="seuratintegrate">SeuratIntegrate<a class="anchor" aria-label="anchor" href="#seuratintegrate"></a>
</h1>
</div>
<p>SeuratIntegrate streamlines single-cell transcriptomics (scRNA-seq) data integration and batch effect correction. This R package effortlessly extends the Seurat workflow with 8 popular integration methods across R and Python, complemented by 11 robust scoring metrics to estimate their performance.</p>
<div class="row">
<div class="col-md-6">
<p><strong>Integrations</strong></p>
<ol style="list-style-type: decimal">
<li>ComBat<br>
</li>
<li>Harmony</li>
<li>MNN</li>
<li>BBKNN</li>
<li>scVI</li>
<li>scANVI</li>
<li>Scanorama</li>
<li>trVAE</li>
</ol>
</div>
<div class="col-md-6">
<p><strong>Scoring metrics</strong></p>
<ol style="list-style-type: decimal">
<li>ARI<br>
</li>
<li>ASW</li>
<li>Batch ASW</li>
<li>Cell cycle conservation</li>
<li>Graph connectivity</li>
<li>PCA-density</li>
<li>PCA-regression</li>
<li>kBET</li>
<li>cell-type (c)LISI</li>
<li>batch (i)LISI</li>
<li>NMI</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install SeuratIntegrate from github:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"remotes"</span>, <span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span> <span class="co"># if not installed</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cbib/Seurat-Integrate"</span>, repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> </span></code></pre></div>
<p>To benefit from SeuratIntegrate’s full capabilities, we recommend installing the following packages:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># fast distance computation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'distances'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># required to test for k-nearest neighbour batch effects</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'theislab/kBET'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># faster Local Inverse Simpson’s Index computation</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'immunogenomics/lisi'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="conda-environments-for-python-methods">Conda environments for Python methods<a class="anchor" aria-label="anchor" href="#conda-environments-for-python-methods"></a>
</h2>
<p>To simplify the creation and management of conda environments, we suggest using <code><a href="reference/UpdateEnvCache.html">UpdateEnvCache()</a></code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create environments:</span></span>
<span><span class="fu"><a href="reference/UpdateEnvCache.html">UpdateEnvCache</a></span><span class="op">(</span><span class="st">"bbknn"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/UpdateEnvCache.html">UpdateEnvCache</a></span><span class="op">(</span><span class="st">"scvi"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/UpdateEnvCache.html">UpdateEnvCache</a></span><span class="op">(</span><span class="st">"scanorama"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/UpdateEnvCache.html">UpdateEnvCache</a></span><span class="op">(</span><span class="st">"trvae"</span><span class="op">)</span></span></code></pre></div>
<p>Those environments will be saved and automatically used by the Python integration methods provided by SeuratIntegrate.</p>
<p>Alternatively, the cache can be updated with a pre-existing environment. This can be useful if you have to set up a conda environment yourself because a command above failed or a conda environment turned out to be non-functional.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save "my_bbknn_env" (for bbknn) to cache</span></span>
<span><span class="fu"><a href="reference/UpdateEnvCache.html">UpdateEnvCache</a></span><span class="op">(</span><span class="st">"bbknn"</span>, conda.env <span class="op">=</span> <span class="st">"my_bbknn_env"</span><span class="op">)</span></span></code></pre></div>
<p>The cache remains persistent across sessions and its state can be displayed with:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/getCache.html">getCache</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="float">
<img src="articles/img/conda_envs_set.png" title="conda envs set" alt="A fully set CondaEnvManager"><div class="figcaption">A fully set <code>CondaEnvManager</code>
</div>
</div>
<p>Further details are provided in the <code><a href="articles/setup_and_tips.html">vignette("setup_and_tips")</a></code>.</p>
</div>
<div class="section level2">
<h2 id="data-integration">Data integration<a class="anchor" aria-label="anchor" href="#data-integration"></a>
</h2>
<p>To run integration algorithms, we have developed a function called <code><a href="reference/DoIntegrate.html">DoIntegrate()</a></code> that enables:</p>
<ul>
<li>performing multiple integration methods in a single call</li>
<li>control over the data (raw, normalised or scaled) and the features to use as input</li>
<li>flexible customisation of parameters for each integration method</li>
</ul>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/DoIntegrate.html">DoIntegrate</a></span><span class="op">(</span><span class="va">seu</span>,</span>
<span>       <span class="co"># ... integrations</span></span>
<span>         <span class="fu"><a href="reference/CombatIntegration.html">CombatIntegration</a></span><span class="op">(</span>layers <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="reference/HarmonyIntegration.html">HarmonyIntegration</a></span><span class="op">(</span>orig <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="reference/ScanoramaIntegration.html">ScanoramaIntegration</a></span><span class="op">(</span>ncores <span class="op">=</span> <span class="fl">4L</span>, layers <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="reference/scVIIntegration.html">scVIIntegration</a></span><span class="op">(</span>layers <span class="op">=</span> <span class="st">"counts"</span>, features <span class="op">=</span> <span class="fu">Features</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">)</span>,</span>
<span>       <span class="co"># ...</span></span>
<span>       use.hvg <span class="op">=</span> <span class="cn">TRUE</span>,    <span class="co"># `VariableFeatures()`</span></span>
<span>       use.future <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Here, all integration methods will use the variable features as input, with the exception of <code><a href="reference/scVIIntegration.html">scVIIntegration()</a></code> which is set to use all features. <code><a href="reference/CombatIntegration.html">CombatIntegration()</a></code> will correct the normalised counts, while <code><a href="reference/scVIIntegration.html">scVIIntegration()</a></code> will train on raw counts.</p>
<p><code>use.future</code> must be <code>TRUE</code> for <strong>Python methods</strong>, and <code>FALSE</code> for <strong>R methods</strong>.</p>
<p>Integration methods produce one or several outputs. Those can be of multiple types - either a new assay with corrected counts, a new dimension reduction with corrected cell embeddings, or a new graph with corrected edges. The type of output is important to consider, because it will require different post-processing steps until the result of the integration can be visualised on a UMAP:</p>
<ul>
<li>Corrected counts: <code>ScaleData()</code> -&gt; <code>RunPCA()</code> -&gt; <code>RunUMAP()</code>
</li>
<li>Dimension reduction: <code>RunUMAP()</code>
</li>
<li>KNN graph: <code>RunUMAP(umap.method = "umap-learn")</code>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="scoring">Scoring<a class="anchor" aria-label="anchor" href="#scoring"></a>
</h2>
<p>SeuratIntegrate incorporates 11 scoring metrics: 6 quantify the degree of batch mixing (<em>batch correction</em>), while 5 assess the preservation of biological differences (<em>bio-conservation</em>) based on ground truth cell type labels.</p>
<p>Each score can be obtained using a function of the form <code>Score[score_name]()</code>, or directly saved in the Seurat object using the <code>AddScore[score_name]()</code> counterpart:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save the score in a variable</span></span>
<span><span class="va">rpca_score</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/score-regressPC.html">ScoreRegressPC</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or save the score in the Seurat object</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/score-regressPC.html">AddScoreRegressPC</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>AddScore</code> functions have an advantage over the <code>Score</code> functions. They allow to then scale the scores between zero and one and to standardize their direction (the closer to one, always the better), improving interpretability. Then, scores can be plotted to readily compare the different integrations:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># scale</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ScaleScores.html">ScaleScores</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="fu"><a href="reference/PlotScores.html">PlotScores</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cbib/Seurat-Integrate/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/cbib/Seurat-Integrate/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing SeuratIntegrate</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Florian Specque <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-4449-6381" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Domitille Chalopin <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-0588-96301" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Florian Specque, Domitille Chalopin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
