<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run Scanorama on Seurat's Assay5 object through IntegrateLayers — ScanoramaIntegration • SeuratIntegrate</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Run Scanorama on Seurat's Assay5 object through IntegrateLayers — ScanoramaIntegration"><meta name="description" content="A wrapper to run Scanorama on multi-layered Seurat V5 object.
Requires a conda environment with scanorama and necessary dependencies"><meta property="og:description" content="A wrapper to run Scanorama on multi-layered Seurat V5 object.
Requires a conda environment with scanorama and necessary dependencies"><meta property="og:image" content="https://fspecque.github.io/test/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SeuratIntegrate</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SeuratIntegrate.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/introduction.html">Integration example: pbmcsca data</a></li>
    <li><a class="dropdown-item" href="../articles/setup_and_tips.html">Advice about conda and future</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Aide-mémoires"><span class="fa fa-sticky-note"></span></button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-"><li><a class="dropdown-item" href="../articles/memo_integration.html">Integration methods</a></li>
    <li><a class="dropdown-item" href="../articles/memo_score.html">Scoring metrics</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cbib/Seurat-Integrate/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Run Scanorama on Seurat's <a href="https://satijalab.github.io/seurat-object/reference/Assay5-class.html" class="external-link">Assay5</a> object through <code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></code></h1>
      <small class="dont-index">Source: <a href="https://github.com/cbib/Seurat-Integrate/blob/main/R/Scanorama.R" class="external-link"><code>R/Scanorama.R</code></a></small>
      <div class="d-none name"><code>ScanoramaIntegration.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A wrapper to run <code>Scanorama</code> on multi-layered Seurat V5 object.
Requires a conda environment with <code>scanorama</code> and necessary dependencies</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ScanoramaIntegration</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">orig</span>,</span>
<span>  layers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  scale.layer <span class="op">=</span> <span class="st">"scale.data"</span>,</span>
<span>  conda_env <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  new.reduction <span class="op">=</span> <span class="st">"integrated.scanorama"</span>,</span>
<span>  reduction.key <span class="op">=</span> <span class="st">"scanorama_"</span>,</span>
<span>  reconstructed.assay <span class="op">=</span> <span class="st">"scanorama.reconstructed"</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  ndims.out <span class="op">=</span> <span class="fl">100L</span>,</span>
<span>  return_dense <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  approx <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sigma <span class="op">=</span> <span class="fl">15L</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  knn <span class="op">=</span> <span class="fl">20L</span>,</span>
<span>  hvg.scanorama <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  union.features <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sketch <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sketch_method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"geosketch"</span>, <span class="st">"uniform"</span><span class="op">)</span>,</span>
<span>  sketch_max <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  seed.use <span class="op">=</span> <span class="fl">42L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>A <code><a href="https://satijalab.github.io/seurat-object/reference/Seurat-class.html" class="external-link">Seurat</a></code> object
(or an <code><a href="https://satijalab.github.io/seurat-object/reference/Assay5-class.html" class="external-link">Assay5</a></code> object if
not called by <code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></code>)</p></dd>


<dt id="arg-orig">orig<a class="anchor" aria-label="anchor" href="#arg-orig"></a></dt>
<dd><p><code><a href="https://satijalab.github.io/seurat-object/reference/DimReduc-class.html" class="external-link">DimReduc</a></code> object. Not to be set
directly when called with <code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></code>, use
<code>orig.reduction</code> argument instead</p></dd>


<dt id="arg-layers">layers<a class="anchor" aria-label="anchor" href="#arg-layers"></a></dt>
<dd><p>Name of the layers to use in the integration</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Vector of feature names to input to the integration method.
When <code>features = NULL</code> (default), the
<code><a href="https://satijalab.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></code> are used. To pass all features,
use the output of <code><a href="https://satijalab.github.io/seurat-object/reference/Cells.html" class="external-link">Features</a>()</code></p></dd>


<dt id="arg-scale-layer">scale.layer<a class="anchor" aria-label="anchor" href="#arg-scale-layer"></a></dt>
<dd><p>Name of the scaled layer in <code>Assay</code></p></dd>


<dt id="arg-conda-env">conda_env<a class="anchor" aria-label="anchor" href="#arg-conda-env"></a></dt>
<dd><p>Path to conda environment to run scanorama (should also
contain the scipy python module). By default, uses the conda environment
registered for scanorama in the conda environment manager</p></dd>


<dt id="arg-new-reduction">new.reduction<a class="anchor" aria-label="anchor" href="#arg-new-reduction"></a></dt>
<dd><p>Name to store resulting
<a href="https://satijalab.github.io/seurat-object/reference/DimReduc-class.html" class="external-link">dimensional reduction</a> object. <code>NULL</code> or
<code>FALSE</code> disables the dimension reduction computation</p></dd>


<dt id="arg-reduction-key">reduction.key<a class="anchor" aria-label="anchor" href="#arg-reduction-key"></a></dt>
<dd><p>Key for resulting
<a href="https://satijalab.github.io/seurat-object/reference/DimReduc-class.html" class="external-link">dimensional reduction</a> object. Ignored if
<code>reduction.key</code> is <code>NULL</code> or <code>FALSE</code></p></dd>


<dt id="arg-reconstructed-assay">reconstructed.assay<a class="anchor" aria-label="anchor" href="#arg-reconstructed-assay"></a></dt>
<dd><p>Name for the <code>assay</code> containing the corrected
expression matrix of dimension #<code>features</code> x #cells. <code>NULL</code> or
<code>FALSE</code> disables the corrected expression matrix computation</p></dd>


<dt id="arg-ncores">ncores<a class="anchor" aria-label="anchor" href="#arg-ncores"></a></dt>
<dd><p>Number of parallel threads to create through
<a href="https://rdrr.io/pkg/RhpcBLASctl/man/RhpcBLASctl-package.html" class="external-link">blas_set_num_threads</a>. Pointless if BLAS is not supporting
multithreaded</p></dd>


<dt id="arg-ndims-out">ndims.out<a class="anchor" aria-label="anchor" href="#arg-ndims-out"></a></dt>
<dd><p>Number of dimensions for <code>new.reduction</code> output.
Scanorama specific argument</p></dd>


<dt id="arg-return-dense">return_dense<a class="anchor" aria-label="anchor" href="#arg-return-dense"></a></dt>
<dd><p>Whether scanorama returns <code>numpy.ndarray</code> matrices
instead of <code>scipy.sparse.csr_matrix</code>. Scanorama specific argument</p></dd>


<dt id="arg-batch-size">batch_size<a class="anchor" aria-label="anchor" href="#arg-batch-size"></a></dt>
<dd><p>Used in the alignment vector computation. Higer values
increase memory burden.Scanorama specific argument</p></dd>


<dt id="arg-approx">approx<a class="anchor" aria-label="anchor" href="#arg-approx"></a></dt>
<dd><p>Whether scanorama approximates nearest neighbors computation.
Scanorama specific argument</p></dd>


<dt id="arg-sigma">sigma<a class="anchor" aria-label="anchor" href="#arg-sigma"></a></dt>
<dd><p>Correction smoothing parameter on gaussian kernel. Scanorama
specific argument</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>Minimum alignment score. Scanorama specific argument</p></dd>


<dt id="arg-knn">knn<a class="anchor" aria-label="anchor" href="#arg-knn"></a></dt>
<dd><p>Number of nearest neighbours used for matching. Scanorama
specific argument</p></dd>


<dt id="arg-hvg-scanorama">hvg.scanorama<a class="anchor" aria-label="anchor" href="#arg-hvg-scanorama"></a></dt>
<dd><p>A positive integer to turn scanorama's internal HVG
selection on. Disabled by default. Scanorama specific argument</p></dd>


<dt id="arg-union-features">union.features<a class="anchor" aria-label="anchor" href="#arg-union-features"></a></dt>
<dd><p>By default, scanorama uses the intersection of features
to perform integration. Set this parameter to <code>TRUE</code> to use the union.
Discouraged. Scanorama specific argument</p></dd>


<dt id="arg-sketch">sketch<a class="anchor" aria-label="anchor" href="#arg-sketch"></a></dt>
<dd><p>Turns on sketching-based acceleration by first downsampling the
datasets. See Hie et al., Cell Systems (2019). Disabled by default. Ignored
when <code>reconstructed.assay</code> is enabled. Scanorama specific argument</p></dd>


<dt id="arg-sketch-method">sketch_method<a class="anchor" aria-label="anchor" href="#arg-sketch-method"></a></dt>
<dd><p>A skething method to apply to the data. Either
'geosketch' (default) or 'uniform'. Ignored when <code>reconstructed.assay</code>
is enabled or when <code>sketch</code> is FALSE. Scanorama specific argument</p></dd>


<dt id="arg-sketch-max">sketch_max<a class="anchor" aria-label="anchor" href="#arg-sketch-max"></a></dt>
<dd><p>Downsampling cutoff. Ignored when sketching disabled.
Scanorama specific argument</p></dd>


<dt id="arg-seed-use">seed.use<a class="anchor" aria-label="anchor" href="#arg-seed-use"></a></dt>
<dd><p>An integer to generate reproducible outputs.
Set <code>seed.use = NULL</code> to disable</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Print messages. Set to <code>FALSE</code> to disable</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Ignored</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing at least one of:</p><ul><li><p>a new DimReduc of name <code>reduction.name</code> (key set to
  <code>reduction.key</code>) with corrected embeddings matrix of <code>ndims.out</code>.</p></li>
<li><p>a new Assay of name <code>reconstructed.assay</code> with corrected counts
  for each <code>features</code> (or less if <code>hvg</code> is set to a lower integer).</p></li>
</ul><p>When called via <code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></code>, a Seurat object with
the new reduction and/or assay is returned</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>This function requires the
<a href="https://github.com/brianhie/scanorama" class="external-link"><span class="pkg">Scanorama</span></a> package
to be installed (along with <span class="pkg">scipy</span>)</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Hie, B., Bryson, B. &amp; Berger, B. Efficient integration of
heterogeneous single-cell transcriptomes using Scanorama. Nat Biotechnol 37,
685–691 (2019). <a href="https://doi.org/10.1038/s41587-019-0113-3" class="external-link">DOI</a></p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://satijalab.github.io/seurat-object/reference/Tool.html" class="external-link">Tool</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Preprocessing</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"pbmcsca"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">obj</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, f <span class="op">=</span> <span class="va">obj</span><span class="op">$</span><span class="va">Method</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># After preprocessing, we integrate layers:</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">IntegrateLayers</span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, method <span class="op">=</span> <span class="va">ScanoramaIntegration</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To disable feature expression matrix correction:</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">IntegrateLayers</span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, method <span class="op">=</span> <span class="va">ScanoramaIntegration</span>,</span></span>
<span class="r-in"><span>                       reconstructed.assay <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Florian Specque, Domitille Chalopin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

