<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run scVI on Seurat's Assay5 object through IntegrateLayers — scVIIntegration • SeuratIntegrate</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Run scVI on Seurat's Assay5 object through IntegrateLayers — scVIIntegration"><meta name="description" content='A wrapper to run scVI on multi-layered Seurat V5 object.
Requires a conda environment with scvi-tools and necessary dependencies
Can be called via SeuratIntegrate::scVIIntegration() or
scVIIntegration.fix()
Recommendations: use raw counts and all features
(features = Features(object), layers = "counts")'><meta property="og:description" content='A wrapper to run scVI on multi-layered Seurat V5 object.
Requires a conda environment with scvi-tools and necessary dependencies
Can be called via SeuratIntegrate::scVIIntegration() or
scVIIntegration.fix()
Recommendations: use raw counts and all features
(features = Features(object), layers = "counts")'><meta property="og:image" content="https://fspecque.github.io/test/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SeuratIntegrate</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/SeuratIntegrate.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/introduction.html">Integration example: pbmcsca data</a></li>
    <li><a class="dropdown-item" href="../articles/setup_and_tips.html">Advice about conda and future</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Aide-mémoires"><span class="fa fa-sticky-note"></span></button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-"><li><a class="dropdown-item" href="../articles/memo_integration.html">Integration methods</a></li>
    <li><a class="dropdown-item" href="../articles/memo_score.html">Scoring metrics</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cbib/Seurat-Integrate/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Run scVI on Seurat's <a href="https://satijalab.github.io/seurat-object/reference/Assay5-class.html" class="external-link">Assay5</a> object through <code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></code></h1>
      <small class="dont-index">Source: <a href="https://github.com/cbib/Seurat-Integrate/blob/main/R/scVI.R" class="external-link"><code>R/scVI.R</code></a></small>
      <div class="d-none name"><code>scVIIntegration.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A wrapper to run <code>scVI</code> on multi-layered Seurat V5 object.
Requires a conda environment with <code>scvi-tools</code> and necessary dependencies</p>
<p>Can be called via <code>SeuratIntegrate::scVIIntegration()</code> or
<code>scVIIntegration.fix()</code></p>
<p><strong>Recommendations</strong>: use raw counts and all features
(<code>features = Features(object), layers = "counts"</code>)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">scVIIntegration</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  groups <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  groups.name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  labels.name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  layers <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  scale.layer <span class="op">=</span> <span class="st">"scale.data"</span>,</span>
<span>  conda_env <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  new.reduction <span class="op">=</span> <span class="st">"integrated.scVI"</span>,</span>
<span>  reduction.key <span class="op">=</span> <span class="st">"scVIlatent_"</span>,</span>
<span>  torch.intraop.threads <span class="op">=</span> <span class="fl">4L</span>,</span>
<span>  torch.interop.threads <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  model.save.dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ndims.out <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  n_hidden <span class="op">=</span> <span class="fl">128L</span>,</span>
<span>  n_layers <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  dropout_rate <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  dispersion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="st">"gene-batch"</span>, <span class="st">"gene-label"</span>, <span class="st">"gene-cell"</span><span class="op">)</span>,</span>
<span>  gene_likelihood <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"zinb"</span>, <span class="st">"nb"</span>, <span class="st">"poisson"</span><span class="op">)</span>,</span>
<span>  latent_distribution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"normal"</span>, <span class="st">"ln"</span><span class="op">)</span>,</span>
<span>  max_epochs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  train_size <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  batch_size <span class="op">=</span> <span class="fl">128L</span>,</span>
<span>  seed.use <span class="op">=</span> <span class="fl">42L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose.scvi <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"INFO"</span>, <span class="st">"NOTSET"</span>, <span class="st">"DEBUG"</span>, <span class="st">"WARNING"</span>, <span class="st">"ERROR"</span>, <span class="st">"CRITICAL"</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">scVIIntegration.fix</span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>A <code><a href="https://satijalab.github.io/seurat-object/reference/Seurat-class.html" class="external-link">Seurat</a></code> object
(or an <code><a href="https://satijalab.github.io/seurat-object/reference/Assay5-class.html" class="external-link">Assay5</a></code> object if
not called by <code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></code>)</p></dd>


<dt id="arg-groups">groups<a class="anchor" aria-label="anchor" href="#arg-groups"></a></dt>
<dd><p>A <b>named</b> data frame with grouping information.</p></dd>


<dt id="arg-groups-name">groups.name<a class="anchor" aria-label="anchor" href="#arg-groups-name"></a></dt>
<dd><p>Column name from <code>groups</code> data frame that stores
grouping information. If <code>groups.name = NULL</code>, the first column is used</p></dd>


<dt id="arg-labels-name">labels.name<a class="anchor" aria-label="anchor" href="#arg-labels-name"></a></dt>
<dd><p>Column name from <code>groups</code> data frame that stores
cell label information. If <code>labels.name = NULL</code>, all cells are assigned
the same label.</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Vector of feature names to input to the integration method.
When <code>features = NULL</code> (default), the
<code><a href="https://satijalab.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></code> are used. To pass all features,
use the output of <code><a href="https://satijalab.github.io/seurat-object/reference/Cells.html" class="external-link">Features</a>()</code></p></dd>


<dt id="arg-layers">layers<a class="anchor" aria-label="anchor" href="#arg-layers"></a></dt>
<dd><p>Name of the layers to use in the integration.
<b>'counts'</b> is highly recommended</p></dd>


<dt id="arg-scale-layer">scale.layer<a class="anchor" aria-label="anchor" href="#arg-scale-layer"></a></dt>
<dd><p>Name of the scaled layer in <code>Assay</code></p></dd>


<dt id="arg-conda-env">conda_env<a class="anchor" aria-label="anchor" href="#arg-conda-env"></a></dt>
<dd><p>Path to conda environment to run scVI (should also
contain the scipy python module).  By default, uses the conda environment
registered for scVI in the conda environment manager</p></dd>


<dt id="arg-new-reduction">new.reduction<a class="anchor" aria-label="anchor" href="#arg-new-reduction"></a></dt>
<dd><p>Name of the new integrated dimensional reduction</p></dd>


<dt id="arg-reduction-key">reduction.key<a class="anchor" aria-label="anchor" href="#arg-reduction-key"></a></dt>
<dd><p>Key for the new integrated dimensional reduction</p></dd>


<dt id="arg-torch-intraop-threads">torch.intraop.threads<a class="anchor" aria-label="anchor" href="#arg-torch-intraop-threads"></a></dt>
<dd><p>Number of intra-op threads available to torch
when training on CPU instead of GPU. Set via <code>torch.set_num_threads()</code>.</p></dd>


<dt id="arg-torch-interop-threads">torch.interop.threads<a class="anchor" aria-label="anchor" href="#arg-torch-interop-threads"></a></dt>
<dd><p>Number of intra-op threads available to torch
when training on CPU instead of GPU. Set via <code>torch.set_num_interop_threads()</code>.
Can only be changed once, on first call.</p></dd>


<dt id="arg-model-save-dir">model.save.dir<a class="anchor" aria-label="anchor" href="#arg-model-save-dir"></a></dt>
<dd><p>Path to a directory to save the model to. Uses
<code>SCVI.save()</code>. Does not save anndata. Note that neither the trainer
optimizer state nor the trainer history are saved.
<code>model.save.dir = NULL</code> (default) disables saving the model.</p></dd>


<dt id="arg-ndims-out">ndims.out<a class="anchor" aria-label="anchor" href="#arg-ndims-out"></a></dt>
<dd><p>Number of dimensions for <code>new.reduction</code> output.
Corresponds to <code>n_latent</code> argument in the original API of SCVI</p></dd>


<dt id="arg-n-hidden">n_hidden<a class="anchor" aria-label="anchor" href="#arg-n-hidden"></a></dt>
<dd><p>Number of nodes per hidden layer.</p></dd>


<dt id="arg-n-layers">n_layers<a class="anchor" aria-label="anchor" href="#arg-n-layers"></a></dt>
<dd><p>Number of hidden layers used for encoder and decoder NNs.</p></dd>


<dt id="arg-dropout-rate">dropout_rate<a class="anchor" aria-label="anchor" href="#arg-dropout-rate"></a></dt>
<dd><p>Dropout rate for neural networks.</p></dd>


<dt id="arg-dispersion">dispersion<a class="anchor" aria-label="anchor" href="#arg-dispersion"></a></dt>
<dd><p>One of the following:</p><ul><li><p><code>gene</code>: dispersion parameter of NB is constant per gene across cells (default)</p></li>
<li><p><code>gene-batch</code>: dispersion can differ between different batches</p></li>
<li><p><code>gene-label</code>: dispersion can differ between different labels</p></li>
<li><p><code>gene-cell</code>: dispersion can differ for every gene in every cell</p></li>
</ul></dd>


<dt id="arg-gene-likelihood">gene_likelihood<a class="anchor" aria-label="anchor" href="#arg-gene-likelihood"></a></dt>
<dd><p>One of the following:</p><ul><li><p><code>zinb</code>: Zero-inflated negative binomial distribution (default)</p></li>
<li><p><code>nb</code>: Negative binomial distribution</p></li>
<li><p><code>poisson</code>: Poisson distribution</p></li>
</ul></dd>


<dt id="arg-latent-distribution">latent_distribution<a class="anchor" aria-label="anchor" href="#arg-latent-distribution"></a></dt>
<dd><p>One of the following:</p><ul><li><p><code>normal</code>: Normal distribution (default)</p></li>
<li><p><code>ln</code>: Logistic normal distribution (Normal(0, I) transformed by softmax)</p></li>
</ul></dd>


<dt id="arg-max-epochs">max_epochs<a class="anchor" aria-label="anchor" href="#arg-max-epochs"></a></dt>
<dd><p>Number of passes through the dataset for semisupervised training.</p></dd>


<dt id="arg-train-size">train_size<a class="anchor" aria-label="anchor" href="#arg-train-size"></a></dt>
<dd><p>Size of training set in the range <code>[0.0, 1.0]</code></p></dd>


<dt id="arg-batch-size">batch_size<a class="anchor" aria-label="anchor" href="#arg-batch-size"></a></dt>
<dd><p>Minibatch size to use during training.</p></dd>


<dt id="arg-seed-use">seed.use<a class="anchor" aria-label="anchor" href="#arg-seed-use"></a></dt>
<dd><p>An integer to generate reproducible outputs.
Set <code>seed.use = NULL</code> to disable</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Print messages. Set to <code>FALSE</code> to disable</p></dd>


<dt id="arg-verbose-scvi">verbose.scvi<a class="anchor" aria-label="anchor" href="#arg-verbose-scvi"></a></dt>
<dd><p>Verbosity level of scVI. From quietest to talkiest:
CRITICAL, ERROR, WARNING, INFO (default), DEBUG, NOTSET</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>For <code>scVIIntegration()</code>, additional arguments to be passed to
<code>scvi.model.SCVI</code>, <code>SCVI.setup_anndata</code> or <code>SCVI.train</code> (see
<strong>Details</strong> section). For <code>scVIIntegration.fix()</code>, all of the above</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing:</p><ul><li><p>a new DimReduc of name <code>new.reduction</code> (key set to
  <code>reduction.key</code>) consisting of the latent space of the model with
  <code>ndims.out</code> dimensions.</p></li>
</ul><p>When called via <code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></code>, a Seurat object with
the new reduction and/or assay is returned</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This wrappers calls three python functions through <span class="pkg">reticulate</span>.
Find the <span class="pkg">scVI</span>-specific arguments there:</p><ul><li><p>model initiation:
  <a href="https://docs.scvi-tools.org/en/stable/api/reference/scvi.model.SCVI.html#scvi.model.SCVI" class="external-link">scvi.model.SCVI</a>, which relies on
  <a href="https://docs.scvi-tools.org/en/stable/api/reference/scvi.module.VAE.html#scvi.module.VAE" class="external-link">scvi.module.VAE</a></p></li>
<li><p>anndata setup:
  <a href="https://docs.scvi-tools.org/en/stable/api/reference/scvi.model.SCVI.html#scvi.model.SCVI.setup_anndata" class="external-link">SCVI.setup_anndata</a></p></li>
<li><p>training:
  <a href="https://docs.scvi-tools.org/en/stable/api/reference/scvi.model.SCVI.html#scvi.model.SCVI.train" class="external-link">SCVI.train</a></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>This function requires the
<a href="https://scvi-tools.org/" class="external-link"><span class="pkg">scvi-tools</span></a> package
to be installed (along with <span class="pkg">scipy</span>)</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Lopez, R., Regier, J., Cole, M. B., Jordan, M. I. &amp; Yosef, N.
Deep generative modeling for single-cell transcriptomics. Nat Methods 15,
1053–1058 (2018). <a href="https://doi.org/10.1038/s41592-018-0229-2" class="external-link">DOI</a></p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></code>, <code><a href="https://satijalab.org/seurat/reference/writing-integration.html" class="external-link">writing-integration</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Preprocessing</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"pbmcsca"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">obj</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, f <span class="op">=</span> <span class="va">obj</span><span class="op">$</span><span class="va">Method</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">NormalizeData</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">ScaleData</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">RunPCA</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># After preprocessing, we integrate layers:</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">IntegrateLayers</span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, method <span class="op">=</span> <span class="va">scVIIntegration</span>,</span></span>
<span class="r-in"><span>                       features <span class="op">=</span> <span class="fu">Features</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span>, conda_env <span class="op">=</span> <span class="st">'scvi-tools'</span>,</span></span>
<span class="r-in"><span>                       layers <span class="op">=</span> <span class="st">'counts'</span>, groups <span class="op">=</span> <span class="va">obj</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span>, groups.name <span class="op">=</span> <span class="st">'Method'</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To enable cell label-guided correction, save the model, add other</span></span></span>
<span class="r-in"><span><span class="co"># 'nuisance' factors and increase number of threads used:</span></span></span>
<span class="r-in"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu">IntegrateLayers</span><span class="op">(</span>object <span class="op">=</span> <span class="va">obj</span>, method <span class="op">=</span> <span class="va">scVIIntegration</span>,</span></span>
<span class="r-in"><span>                       features <span class="op">=</span> <span class="fu">Features</span><span class="op">(</span><span class="va">obj</span><span class="op">)</span>, conda_env <span class="op">=</span> <span class="st">'scvi-tools'</span>,</span></span>
<span class="r-in"><span>                       layers <span class="op">=</span> <span class="st">'counts'</span>, groups <span class="op">=</span> <span class="va">obj</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span>, groups.name <span class="op">=</span> <span class="st">"Method"</span>,</span></span>
<span class="r-in"><span>                       labels.name <span class="op">=</span> <span class="st">"CellType"</span>,</span></span>
<span class="r-in"><span>                       categorical_covariate_keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Experiment"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                       continuous_covariate_keys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"percent.mito"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                       ncores <span class="op">=</span> <span class="fl">8</span>, model.save.dir <span class="op">=</span> <span class="st">'~/Documents/scVI.model'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Florian Specque, Domitille Chalopin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

