<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Integration example with of the pbmcsca data • SeuratIntegrate</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Integration example with of the pbmcsca data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SeuratIntegrate</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/SeuratIntegrate.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introduction.html">Integration example: pbmcsca data</a></li>
    <li><a class="dropdown-item" href="../articles/setup_and_tips.html">Advice about conda and future</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Aide-mémoires"><span class="fa fa-sticky-note"></span></button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-">
<li><a class="dropdown-item" href="../articles/memo_integration.html">Integration methods</a></li>
    <li><a class="dropdown-item" href="../articles/memo_score.html">Scoring metrics</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cbib/Seurat-Integrate/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="introduction_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="introduction_files/pagedtable-1.1/js/pagedtable.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Integration example with of the pbmcsca data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cbib/Seurat-Integrate/blob/main/vignettes/introduction.Rmd" class="external-link"><code>vignettes/introduction.Rmd</code></a></small>
      <div class="d-none name"><code>introduction.Rmd</code></div>
    </div>

    
    
<style type="text/css">
.darker-code {
  color: #2f2f2f;
}
</style>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cbib/Seurat-Integrate" class="external-link">SeuratIntegrate</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="install-and-load-the-data">Install and load the data<a class="anchor" aria-label="anchor" href="#install-and-load-the-data"></a>
</h2>
<p>We will use the <code>pbmcsca</code> dataset available from the
package <code>SeuratData</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install `SeuratData` package (if not yet)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"SeuratData"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">'satijalab/seurat-data'</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># increase download timeout and install the dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout <span class="op">=</span> <span class="fl">300</span><span class="op">)</span></span>
<span><span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">'pbmcsca'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the dataset (take 1,000 first cells to speed-up execution)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">SeuratData</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">'pbmcsca'</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">1e3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Assay RNA changing from Assay to Assay</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Assay RNA changing from Assay to Assay5</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="inspect-the-dataset">Inspect the dataset<a class="anchor" aria-label="anchor" href="#inspect-the-dataset"></a>
</h2>
<p>Have a look at the metadata:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># rmarkdown::paged_table -&gt; prints data frames in a less ugly way than default</span></span>
<span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/paged_table.html" class="external-link">paged_table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">seu</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["orig.ident"],"name":[1],"type":["fct"],"align":["left"]},{"label":["nCount_RNA"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["nFeature_RNA"],"name":[3],"type":["int"],"align":["right"]},{"label":["nGene"],"name":[4],"type":["chr"],"align":["left"]},{"label":["nUMI"],"name":[5],"type":["chr"],"align":["left"]},{"label":["percent.mito"],"name":[6],"type":["chr"],"align":["left"]},{"label":["Cluster"],"name":[7],"type":["chr"],"align":["left"]},{"label":["CellType"],"name":[8],"type":["chr"],"align":["left"]},{"label":["Experiment"],"name":[9],"type":["chr"],"align":["left"]},{"label":["Method"],"name":[10],"type":["chr"],"align":["left"]}],"data":[{"1":"pbmc1","2":"437125","3":"2200","4":"2200","5":"437125","6":"0.0297434465355702","7":"0","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_108"},{"1":"pbmc1","2":"335596","3":"2438","4":"2438","5":"335596","6":"0.0311521658159055","7":"0","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_115"},{"1":"pbmc1","2":"302204","3":"1874","4":"1874","5":"302204","6":"0.0431128105727693","7":"0","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_133"},{"1":"pbmc1","2":"377420","3":"2480","4":"2480","5":"377420","6":"0.0260323569927476","7":"0","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_142"},{"1":"pbmc1","2":"385514","3":"2196","4":"2196","5":"385514","6":"0.0404759383962183","7":"0","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_143"},{"1":"pbmc1","2":"304994","3":"2216","4":"2216","5":"304994","6":"0.023409951391094","7":"0","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_144"},{"1":"pbmc1","2":"410522","3":"2331","4":"2331","5":"410522","6":"0.0266065031418581","7":"0","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_146"},{"1":"pbmc1","2":"335187","3":"2075","4":"2075","5":"335187","6":"0.0446061115049967","7":"0","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_148"},{"1":"pbmc1","2":"485530","3":"2830","4":"2830","5":"485530","6":"0.0234517593805909","7":"5","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_152"},{"1":"pbmc1","2":"280213","3":"2263","4":"2263","5":"280213","6":"0.0346036948333838","7":"0","8":"Cytotoxic T cell","9":"pbmc1","10":"Smart-seq2","_rn_":"pbmc1_SM2_Cell_153"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>The column <code>Method</code> provides information about the
single-cell technology used to acquire each cell’s trancriptome. We
consider it as the batch of origin.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">Method</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## 10x Chromium (v2) A            CEL-Seq2          Smart-seq2 </span></span>
<span><span class="co">##                 494                 253                 253</span></span></code></pre>
<p>The column <code>CellType</code> contains cell-types annotation that
we will use later on.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/paged_table.html" class="external-link">paged_table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">seu</span><span class="op">$</span><span class="va">CellType</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Var1"],"name":[1],"type":["fct"],"align":["left"]},{"label":["Freq"],"name":[2],"type":["int"],"align":["right"]}],"data":[{"1":"B cell","2":"129"},{"1":"CD14+ monocyte","2":"423"},{"1":"CD16+ monocyte","2":"30"},{"1":"CD4+ T cell","2":"125"},{"1":"Cytotoxic T cell","2":"221"},{"1":"Dendritic cell","2":"4"},{"1":"Megakaryocyte","2":"32"},{"1":"Natural killer cell","2":"10"},{"1":"Plasmacytoid dendritic cell","2":"26"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Let’s now define the batch variable and the cell-type
variable(s):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batch.var</span> <span class="op">&lt;-</span> <span class="st">'Method'</span>   <span class="co"># available in the metadata of the object</span></span>
<span><span class="va">cell.var</span> <span class="op">&lt;-</span> <span class="st">'CellType'</span>  <span class="co"># available in the metadata of the object</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="seurat-data-processing---standard-workflow">Seurat data processing - Standard workflow<a class="anchor" aria-label="anchor" href="#seurat-data-processing---standard-workflow"></a>
</h2>
<p>The current Seurat object is not split, meaning that each layer
contains cells from all batches.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'Layers before split:'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu">Layers</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, <span class="st">'\n'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Layers before split: counts, data</span></span></code></pre>
<p>We select the column <code>Method</code> to separate the cells into
batch-specific layers. This step is indispensable to run integration
methods and enables to normalise each batch (i.e. layer)
independently.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">seu</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span>, f <span class="op">=</span> <span class="va">seu</span><span class="op">$</span><span class="va">Method</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'Layers after split:'</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu">Layers</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, <span class="st">'\n'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Layers after split: counts.Smart-seq2, counts.CEL-Seq2, counts.10x_Chromium_v2_A, data.Smart-seq2, data.CEL-Seq2, data.10x_Chromium_v2_A</span></span></code></pre>
<p>Then, we proceed to the standard <code>Seurat</code> workflow until
we obtain the PCA reduction.</p>
<p><a id="standard-worflow"></a></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seu</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, k.param <span class="op">=</span> <span class="fl">20L</span><span class="op">)</span></span></code></pre></div>
<p>We now further process the object until we can visualise the
dispersion of cells on a UMAP dimension reduction.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">'SCT_snn'</span>, resolution <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seu</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, reduction <span class="op">=</span> <span class="st">'pca'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 1000</span></span>
<span><span class="co">## Number of edges: 43767</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.8715</span></span>
<span><span class="co">## Number of communities: 8</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html" class="external-link">NoLegend</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span>ratio <span class="op">=</span> <span class="fl">.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-8-1.png" alt="UMAP of unintegrated pbmcsca data with cells colored by cluster" width="700"></p>
<p>Let’s colour cells according to batch and cell-type label.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, group.by <span class="op">=</span> <span class="va">batch.var</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, group.by <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span>ratio <span class="op">=</span> <span class="fl">.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-9-1.png" alt="UMAP of unintegrated pbmcsca data with cells colored by sequencing method (i.e. batch) and cell type label" width="1200"></p>
<p>Albeit moderately, the dispersion of cells on the UMAP seems to be
influenced by a batch effect. According to the “CellType” variable, B
cells from 10x and CEL-Seq2 are not gathered together and monocyte cells
do not fully overlap.</p>
<p>Now, we want to correct these technical differences.</p>
</div>
<div class="section level2">
<h2 id="integrate-batches---seuratintegrate-workflow">Integrate batches - SeuratIntegrate workflow<a class="anchor" aria-label="anchor" href="#integrate-batches---seuratintegrate-workflow"></a>
</h2>
<p>The integration commands are to some extent similar to the <a href="https://satijalab.org/seurat/articles/seurat5_integration" class="external-link">Seurat
V5 vignette</a>. The purpose of this package is to extend the set of
available integration methods. See the bottom of
<code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">?IntegrateLayers</a></code> to have a comprehensive list of relevant
methods.</p>
<p>Many methods supported by SeuratIntegrate are implemented in Python,
and the wrappers provided rely on the <code>reticulate</code> package
and <code>conda</code> environments. If you are not familiar with the
<code>CondaEnvManager</code>, have a look at the
<code><a href="../articles/setup_and_tips.html">vignette("setup_and_tips")</a></code>.</p>
<p>Here are some important considerations before performing an
integration:</p>
<ul>
<li>use the interface <code><a href="../reference/DoIntegrate.html">DoIntegrate()</a></code> from SeuratIntegrate to
integrate the cell batches (see the next <a href="#id_#do_integrate">section</a>)</li>
<li>check the method-specific arguments
(e.g. <code><a href="../reference/bbknnIntegration.html">?bbknnIntegration</a></code>)</li>
<li>set up conda environments to use Python-based methods:
<ul>
<li>with the <code>CondaEnvManager</code> via
<code><a href="../reference/UpdateEnvCache.html">UpdateEnvCache()</a></code> (check the
<code><a href="../articles/setup_and_tips.html">vignette("setup_and_tips")</a></code>)</li>
<li>with a custom conda environment by passing its path - or its name -
to the <code>conda_env</code> parameter. This overrides the default
behaviour by loading the specified environment instead of fetching the
<code>CondaEnvManager</code>’s cache.</li>
</ul>
</li>
</ul>
<p>Some methods expect a raw count matrix, others expect the scaled
counts of the variable features, etc. To help you with the choice, look
at the table below</p>
<table style="width:99%;" class="table">
<colgroup>
<col width="14%">
<col width="23%">
<col width="61%">
</colgroup>
<thead><tr class="header">
<th>Method</th>
<th>Layer</th>
<th>Features</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>ComBat</td>
<td>data</td>
<td>any but better with restricted number of features
(e.g. variable)</td>
</tr>
<tr class="even">
<td>Harmony</td>
<td>N/A (PCA reduction)</td>
<td>N/A (PCA reduction)</td>
</tr>
<tr class="odd">
<td>MNN</td>
<td>data</td>
<td>any but better with restricted number of features
(e.g. variable)</td>
</tr>
<tr class="even">
<td>bbknn</td>
<td>N/A (PCA reduction)</td>
<td>N/A (PCA reduction)</td>
</tr>
<tr class="odd">
<td>scVI / scANVI</td>
<td>counts</td>
<td>all</td>
</tr>
<tr class="even">
<td>Scanorama</td>
<td>counts or data</td>
<td>any</td>
</tr>
<tr class="odd">
<td>trVAE</td>
<td>
<p>data (recon loss “mse”)</p>
<p>counts (otherwise)</p>
</td>
<td>all</td>
</tr>
</tbody>
</table>
<p>Layers:</p>
<ul>
<li>counts: raw counts</li>
<li>data: normalised counts</li>
<li>scale.data: scaled normalised counts of variable features</li>
</ul>
<p><br><strong style="color:red;"> /!\ IMPORTANT /!\ </strong> To use
all features when calling an integration method with
<code><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers()</a></code>:
<code>IntegrateLayers(object, features = Features(object), scale.layer = NULL)</code>.
Does not work for a <code>SCTAssay</code>.</p>
<div class="section level3">
<h3 id="dointegrate-philosophy-do_integrate">DoIntegrate philosophy (#do_integrate)<a class="anchor" aria-label="anchor" href="#dointegrate-philosophy-do_integrate"></a>
</h3>
<p>The function <code><a href="../reference/DoIntegrate.html">DoIntegrate()</a></code> is a very handy way to run
multiple integrations in a single command. Note that:</p>
<ul>
<li>
<code>...</code> is the place to specify the integration commands to
run.</li>
<li>integration commands are expected to be <strong>function
calls</strong>, i.e. of the form <code>FooIntegration()</code>,
<code>BarIntegration()</code>, etc.</li>
<li>Calls accept method-specific argument (e.g.
<code>FooIntegration(layers = "data")</code>)</li>
<li>
<code>use.hvg = TRUE</code> results in using variable features</li>
<li>
<code>use.future = TRUE</code> is useful to run Python-based methods
(a normal R session cannot load more than one conda instance, while
<code>future</code> enables to launch background sessions, preventing
the main on to load any conda environment.). <strong>It is highly
recommended to set it to <code>FALSE</code> for R-based
methods</strong>.</li>
</ul>
<p>Most integration methods can be used without modifying default
parameters. In this vignette, we will change some arguments to meet our
needs. Notably, we will change the number of cores allocated to each
method (where possible).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ncores</span> <span class="op">&lt;-</span> <span class="fu">parallel</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/parallel/detectCores.html" class="external-link">detectCores</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">2</span></span></code></pre></div>
<p>In this vignette, we are going to use 3 Python-based methods, namely
<a href="https://github.com/Teichlab/bbknn" class="external-link">BBKNN</a>, <a href="https://github.com/brianhie/scanorama" class="external-link">Scanorama</a> and <a href="https://github.com/scverse/scvi-tools" class="external-link">scANVI</a> from the
scvi-tools suite. Let’s make sure they are available straight away:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># BBKNN</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="fu"><a href="../reference/isValid.html">isValid</a></span><span class="op">(</span><span class="fu"><a href="../reference/getCache.html">getCache</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">bbknn</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/UpdateEnvCache.html">UpdateEnvCache</a></span><span class="op">(</span><span class="st">"bbknn"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># Scanorama</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="fu"><a href="../reference/isValid.html">isValid</a></span><span class="op">(</span><span class="fu"><a href="../reference/getCache.html">getCache</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">scanorama</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/UpdateEnvCache.html">UpdateEnvCache</a></span><span class="op">(</span><span class="st">"scanorama"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># scvi-tools</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="fu"><a href="../reference/isValid.html">isValid</a></span><span class="op">(</span><span class="fu"><a href="../reference/getCache.html">getCache</a></span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="va">scanvi</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/UpdateEnvCache.html">UpdateEnvCache</a></span><span class="op">(</span><span class="st">"scanvi"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><br> Let’s proceed to a few batch-effect corrections:</p>
</div>
<div class="section level3">
<h3 id="integration-part-1-using-r-based-methods">Integration part 1) Using R-based methods<a class="anchor" aria-label="anchor" href="#integration-part-1-using-r-based-methods"></a>
</h3>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DoIntegrate.html">DoIntegrate</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">seu</span>,</span>
<span>  <span class="fu">SeuratIntegrate</span><span class="fu">::</span><span class="fu"><a href="../reference/HarmonyIntegration.html">HarmonyIntegration</a></span><span class="op">(</span>orig <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>                                      ncores <span class="op">=</span> <span class="va">ncores</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/CCAIntegration.html" class="external-link">CCAIntegration</a></span><span class="op">(</span>orig <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span> , new.reduction <span class="op">=</span> <span class="st">"cca.integrated"</span>,</span>
<span>                 normalization.method <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RPCAIntegration.html" class="external-link">RPCAIntegration</a></span><span class="op">(</span>orig <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, new.reduction <span class="op">=</span> <span class="st">"rpca.integrated"</span>,</span>
<span>                  normalization.method <span class="op">=</span> <span class="st">"SCT"</span><span class="op">)</span>,</span>
<span>  use.future <span class="op">=</span> <span class="cn">FALSE</span>  <span class="co"># R-based</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Integration 1 in 3: integrating using 'SeuratIntegrate::HarmonyIntegration'</span></span>
<span><span class="co">## Integration 2 in 3: integrating using 'CCAIntegration'</span></span>
<span><span class="co">## Integration 3 in 3: integrating using 'RPCAIntegration'</span></span></code></pre>
<p><em>Note</em>: We use <code>SeuratIntegrate::</code> before
<code>HarmonyIntegration</code> to avoid any confusion with
<code><a href="https://satijalab.org/seurat/reference/HarmonyIntegration.html" class="external-link">Seurat::HarmonyIntegration()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="integration-part-2-using-python-based-methods">Integration part 2) Using python-based methods<a class="anchor" aria-label="anchor" href="#integration-part-2-using-python-based-methods"></a>
</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DoIntegrate.html">DoIntegrate</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">seu</span>,</span>
<span>  <span class="fu"><a href="../reference/bbknnIntegration.html">bbknnIntegration</a></span><span class="op">(</span>orig <span class="op">=</span> <span class="st">"pca"</span>, layers <span class="op">=</span> <span class="st">"data"</span>, ndims <span class="op">=</span> <span class="fl">30</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/ScanoramaIntegration.html">ScanoramaIntegration</a></span><span class="op">(</span>orig <span class="op">=</span> <span class="st">"pca"</span>, ncores <span class="op">=</span> <span class="va">ncores</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="../reference/scANVIIntegration.html">scANVIIntegration</a></span><span class="op">(</span>groups <span class="op">=</span> <span class="va">seu</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span>, groups.name <span class="op">=</span> <span class="st">"Method"</span>,</span>
<span>                    labels.name <span class="op">=</span> <span class="st">"CellType"</span>, layers <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                    torch.intraop.threads <span class="op">=</span> <span class="va">ncores</span>,</span>
<span>                    torch.interop.threads <span class="op">=</span> <span class="va">ncores</span>,</span>
<span>                    max_epochs <span class="op">=</span> <span class="fl">20L</span><span class="op">)</span>,</span>
<span>  use.future <span class="op">=</span> <span class="cn">TRUE</span>,  <span class="co"># Python-based</span></span>
<span>  use.hvg <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Integration 1 in 3: integrating using 'bbknnIntegration'</span></span>
<span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 1000</span></span>
<span><span class="co">## Number of edges: 6866</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.6812</span></span>
<span><span class="co">## Number of communities: 10</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span>
<span><span class="co">## Integration 2 in 3: integrating using 'ScanoramaIntegration'</span></span>
<span><span class="co">## Found 3000 genes among all datasets</span></span>
<span><span class="co">## Processing datasets (0, 1)</span></span>
<span><span class="co">## Processing datasets (0, 2)</span></span>
<span><span class="co">## Processing datasets (1, 2)</span></span>
<span><span class="co">## Integration 3 in 3: integrating using 'scANVIIntegration'</span></span>
<span><span class="co">## <span style="color: #0000BB;">INFO    </span> Training for <span style="color: #00BBBB; font-weight: bold;">20</span> epochs.                                                </span></span>
<span><span class="co">## Training:   0%|          | 0/20 [00:00&lt;?, ?it/s]Epoch 1/20:   0%|          | 0/20 [00:00&lt;?, ?it/s]Epoch 1/20:   5%|▌         | 1/20 [00:01&lt;00:35,  1.89s/it]Epoch 1/20:   5%|▌         | 1/20 [00:01&lt;00:35,  1.89s/it, v_num=1, train_loss_step=1.45e+4, train_loss_epoch=1.69e+4]Epoch 2/20:   5%|▌         | 1/20 [00:01&lt;00:35,  1.89s/it, v_num=1, train_loss_step=1.45e+4, train_loss_epoch=1.69e+4]Epoch 2/20:  10%|█         | 2/20 [00:03&lt;00:32,  1.82s/it, v_num=1, train_loss_step=1.45e+4, train_loss_epoch=1.69e+4]Epoch 2/20:  10%|█         | 2/20 [00:03&lt;00:32,  1.82s/it, v_num=1, train_loss_step=1.38e+4, train_loss_epoch=1.37e+4]Epoch 3/20:  10%|█         | 2/20 [00:03&lt;00:32,  1.82s/it, v_num=1, train_loss_step=1.38e+4, train_loss_epoch=1.37e+4]Epoch 3/20:  15%|█▌        | 3/20 [00:05&lt;00:30,  1.78s/it, v_num=1, train_loss_step=1.38e+4, train_loss_epoch=1.37e+4]Epoch 3/20:  15%|█▌        | 3/20 [00:05&lt;00:30,  1.78s/it, v_num=1, train_loss_step=1.07e+4, train_loss_epoch=1.18e+4]Epoch 4/20:  15%|█▌        | 3/20 [00:05&lt;00:30,  1.78s/it, v_num=1, train_loss_step=1.07e+4, train_loss_epoch=1.18e+4]Epoch 4/20:  20%|██        | 4/20 [00:07&lt;00:28,  1.78s/it, v_num=1, train_loss_step=1.07e+4, train_loss_epoch=1.18e+4]Epoch 4/20:  20%|██        | 4/20 [00:07&lt;00:28,  1.78s/it, v_num=1, train_loss_step=1.66e+4, train_loss_epoch=1.08e+4]Epoch 5/20:  20%|██        | 4/20 [00:07&lt;00:28,  1.78s/it, v_num=1, train_loss_step=1.66e+4, train_loss_epoch=1.08e+4]Epoch 5/20:  25%|██▌       | 5/20 [00:08&lt;00:26,  1.75s/it, v_num=1, train_loss_step=1.66e+4, train_loss_epoch=1.08e+4]Epoch 5/20:  25%|██▌       | 5/20 [00:08&lt;00:26,  1.75s/it, v_num=1, train_loss_step=9.38e+3, train_loss_epoch=1.03e+4]Epoch 6/20:  25%|██▌       | 5/20 [00:08&lt;00:26,  1.75s/it, v_num=1, train_loss_step=9.38e+3, train_loss_epoch=1.03e+4]Epoch 6/20:  30%|███       | 6/20 [00:10&lt;00:24,  1.72s/it, v_num=1, train_loss_step=9.38e+3, train_loss_epoch=1.03e+4]Epoch 6/20:  30%|███       | 6/20 [00:10&lt;00:24,  1.72s/it, v_num=1, train_loss_step=1.72e+4, train_loss_epoch=9.9e+3] Epoch 7/20:  30%|███       | 6/20 [00:10&lt;00:24,  1.72s/it, v_num=1, train_loss_step=1.72e+4, train_loss_epoch=9.9e+3]Epoch 7/20:  35%|███▌      | 7/20 [00:12&lt;00:22,  1.71s/it, v_num=1, train_loss_step=1.72e+4, train_loss_epoch=9.9e+3]Epoch 7/20:  35%|███▌      | 7/20 [00:12&lt;00:22,  1.71s/it, v_num=1, train_loss_step=1.05e+4, train_loss_epoch=9.68e+3]Epoch 8/20:  35%|███▌      | 7/20 [00:12&lt;00:22,  1.71s/it, v_num=1, train_loss_step=1.05e+4, train_loss_epoch=9.68e+3]Epoch 8/20:  40%|████      | 8/20 [00:13&lt;00:20,  1.70s/it, v_num=1, train_loss_step=1.05e+4, train_loss_epoch=9.68e+3]Epoch 8/20:  40%|████      | 8/20 [00:13&lt;00:20,  1.70s/it, v_num=1, train_loss_step=1.78e+4, train_loss_epoch=9.5e+3] Epoch 9/20:  40%|████      | 8/20 [00:13&lt;00:20,  1.70s/it, v_num=1, train_loss_step=1.78e+4, train_loss_epoch=9.5e+3]Epoch 9/20:  45%|████▌     | 9/20 [00:15&lt;00:18,  1.69s/it, v_num=1, train_loss_step=1.78e+4, train_loss_epoch=9.5e+3]Epoch 9/20:  45%|████▌     | 9/20 [00:15&lt;00:18,  1.69s/it, v_num=1, train_loss_step=7.85e+3, train_loss_epoch=9.37e+3]Epoch 10/20:  45%|████▌     | 9/20 [00:15&lt;00:18,  1.69s/it, v_num=1, train_loss_step=7.85e+3, train_loss_epoch=9.37e+3]Epoch 10/20:  50%|█████     | 10/20 [00:17&lt;00:16,  1.70s/it, v_num=1, train_loss_step=7.85e+3, train_loss_epoch=9.37e+3]Epoch 10/20:  50%|█████     | 10/20 [00:17&lt;00:16,  1.70s/it, v_num=1, train_loss_step=7.37e+3, train_loss_epoch=9.26e+3]Epoch 11/20:  50%|█████     | 10/20 [00:17&lt;00:16,  1.70s/it, v_num=1, train_loss_step=7.37e+3, train_loss_epoch=9.26e+3]Epoch 11/20:  55%|█████▌    | 11/20 [00:18&lt;00:15,  1.70s/it, v_num=1, train_loss_step=7.37e+3, train_loss_epoch=9.26e+3]Epoch 11/20:  55%|█████▌    | 11/20 [00:18&lt;00:15,  1.70s/it, v_num=1, train_loss_step=9.05e+3, train_loss_epoch=9.18e+3]Epoch 12/20:  55%|█████▌    | 11/20 [00:18&lt;00:15,  1.70s/it, v_num=1, train_loss_step=9.05e+3, train_loss_epoch=9.18e+3]Epoch 12/20:  60%|██████    | 12/20 [00:20&lt;00:13,  1.70s/it, v_num=1, train_loss_step=9.05e+3, train_loss_epoch=9.18e+3]Epoch 12/20:  60%|██████    | 12/20 [00:20&lt;00:13,  1.70s/it, v_num=1, train_loss_step=9.66e+3, train_loss_epoch=9.11e+3]Epoch 13/20:  60%|██████    | 12/20 [00:20&lt;00:13,  1.70s/it, v_num=1, train_loss_step=9.66e+3, train_loss_epoch=9.11e+3]Epoch 13/20:  65%|██████▌   | 13/20 [00:22&lt;00:11,  1.69s/it, v_num=1, train_loss_step=9.66e+3, train_loss_epoch=9.11e+3]Epoch 13/20:  65%|██████▌   | 13/20 [00:22&lt;00:11,  1.69s/it, v_num=1, train_loss_step=3.98e+3, train_loss_epoch=9.06e+3]Epoch 14/20:  65%|██████▌   | 13/20 [00:22&lt;00:11,  1.69s/it, v_num=1, train_loss_step=3.98e+3, train_loss_epoch=9.06e+3]Epoch 14/20:  70%|███████   | 14/20 [00:24&lt;00:10,  1.68s/it, v_num=1, train_loss_step=3.98e+3, train_loss_epoch=9.06e+3]Epoch 14/20:  70%|███████   | 14/20 [00:24&lt;00:10,  1.68s/it, v_num=1, train_loss_step=1.16e+4, train_loss_epoch=8.98e+3]Epoch 15/20:  70%|███████   | 14/20 [00:24&lt;00:10,  1.68s/it, v_num=1, train_loss_step=1.16e+4, train_loss_epoch=8.98e+3]Epoch 15/20:  75%|███████▌  | 15/20 [00:25&lt;00:08,  1.67s/it, v_num=1, train_loss_step=1.16e+4, train_loss_epoch=8.98e+3]Epoch 15/20:  75%|███████▌  | 15/20 [00:25&lt;00:08,  1.67s/it, v_num=1, train_loss_step=1.49e+4, train_loss_epoch=8.9e+3] Epoch 16/20:  75%|███████▌  | 15/20 [00:25&lt;00:08,  1.67s/it, v_num=1, train_loss_step=1.49e+4, train_loss_epoch=8.9e+3]Epoch 16/20:  80%|████████  | 16/20 [00:27&lt;00:06,  1.66s/it, v_num=1, train_loss_step=1.49e+4, train_loss_epoch=8.9e+3]Epoch 16/20:  80%|████████  | 16/20 [00:27&lt;00:06,  1.66s/it, v_num=1, train_loss_step=4.98e+3, train_loss_epoch=8.86e+3]Epoch 17/20:  80%|████████  | 16/20 [00:27&lt;00:06,  1.66s/it, v_num=1, train_loss_step=4.98e+3, train_loss_epoch=8.86e+3]Epoch 17/20:  85%|████████▌ | 17/20 [00:28&lt;00:04,  1.66s/it, v_num=1, train_loss_step=4.98e+3, train_loss_epoch=8.86e+3]Epoch 17/20:  85%|████████▌ | 17/20 [00:28&lt;00:04,  1.66s/it, v_num=1, train_loss_step=6.78e+3, train_loss_epoch=8.81e+3]Epoch 18/20:  85%|████████▌ | 17/20 [00:28&lt;00:04,  1.66s/it, v_num=1, train_loss_step=6.78e+3, train_loss_epoch=8.81e+3]Epoch 18/20:  90%|█████████ | 18/20 [00:30&lt;00:03,  1.66s/it, v_num=1, train_loss_step=6.78e+3, train_loss_epoch=8.81e+3]Epoch 18/20:  90%|█████████ | 18/20 [00:30&lt;00:03,  1.66s/it, v_num=1, train_loss_step=5.82e+3, train_loss_epoch=8.79e+3]Epoch 19/20:  90%|█████████ | 18/20 [00:30&lt;00:03,  1.66s/it, v_num=1, train_loss_step=5.82e+3, train_loss_epoch=8.79e+3]Epoch 19/20:  95%|█████████▌| 19/20 [00:32&lt;00:01,  1.68s/it, v_num=1, train_loss_step=5.82e+3, train_loss_epoch=8.79e+3]Epoch 19/20:  95%|█████████▌| 19/20 [00:32&lt;00:01,  1.68s/it, v_num=1, train_loss_step=5.13e+3, train_loss_epoch=8.72e+3]Epoch 20/20:  95%|█████████▌| 19/20 [00:32&lt;00:01,  1.68s/it, v_num=1, train_loss_step=5.13e+3, train_loss_epoch=8.72e+3]Epoch 20/20: 100%|██████████| 20/20 [00:34&lt;00:00,  1.68s/it, v_num=1, train_loss_step=5.13e+3, train_loss_epoch=8.72e+3]Epoch 20/20: 100%|██████████| 20/20 [00:34&lt;00:00,  1.68s/it, v_num=1, train_loss_step=1.44e+4, train_loss_epoch=8.68e+3]Epoch 20/20: 100%|██████████| 20/20 [00:34&lt;00:00,  1.70s/it, v_num=1, train_loss_step=1.44e+4, train_loss_epoch=8.68e+3]</span></span></code></pre>
<blockquote>
<p>Note: set <code>max_epochs = 20L</code> for
<code>scANVIIntegration</code> is only to save time ! The default number
of epochs (400) results in a superior integration.</p>
</blockquote>
<p>If we take a look at our Seurat object, we can note that it has been
enriched with many objects:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 69594 features across 1000 samples within 4 assays </span></span>
<span><span class="co">## Active assay: SCT (16450 features, 3000 variable features)</span></span>
<span><span class="co">##  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">##  3 other assays present: RNA, bbknn.ridge, scanorama.reconstructed</span></span>
<span><span class="co">##  8 dimensional reductions calculated: pca, umap, harmony, cca.integrated, rpca.integrated, pca.bbknn, integrated.scanorama, integrated.scANVI</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Graph objects:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu">Graphs</span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Neighbor objects:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/ObjectAccess.html" class="external-link">Neighbors</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Reduction dimensions:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/ObjectAccess.html" class="external-link">Reductions</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Assays:"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/ObjectAccess.html" class="external-link">Assays</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">", "</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Graph objects: SCT_nn, SCT_snn, bbknn_scale.data_connectivities, bbknn_scale.data_distances, bbknn_ridge.residuals_connectivities, bbknn_ridge.residuals_distances </span></span>
<span><span class="co">## Neighbor objects:  </span></span>
<span><span class="co">## Reduction dimensions: pca, umap, harmony, cca.integrated, rpca.integrated, pca.bbknn, integrated.scanorama, integrated.scANVI </span></span>
<span><span class="co">## Assays: RNA, SCT, bbknn.ridge, scanorama.reconstructed</span></span></code></pre>
<p>Great! We have successfully performed several integrations! However,
stopping here would be unsatisfactory because we still need to process
each integration’s output(s) to obtain at least one UMAP projection for
each. Here, we will also aim at generating assessable representations to
score.</p>
</div>
</div>
<div class="section level2">
<h2 id="score-integrations">Score integrations<a class="anchor" aria-label="anchor" href="#score-integrations"></a>
</h2>
<div class="section level3">
<h3 id="process-outputs">Process outputs<a class="anchor" aria-label="anchor" href="#process-outputs"></a>
</h3>
<p>Several objects can be produced by each integration algorithm, namely
a layer in a new assay (i.e. corrected counts), a dimension reduction
(corrected embedding), or a knn network. Some even produce more than one
output (for instance Scanorama produces corrected counts and a dimension
reduction).</p>
<p>The type of output is important to consider, because scoring metrics
are not compatible with all output types. The simplest strategy is to
process each output separately in order to obtain at least a PCA out of
it, or even a knn graph (essential to compute clusters). Note that most
scores cannot be computed on knn graphs, hence knn graph outputs
(e.g. BBKNN) can only be evaluated by a reduced set of metrics.</p>
<p>Below is a summary of post-processing steps for each output type
(bracketed steps are not always necessary):</p>
<ul>
<li>corrected counts: [<code><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData()</a></code>] -&gt;
<code><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA()</a></code> -&gt; [<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors()</a></code> -&gt;
<code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code>]</li>
<li>dimension reduction: [<code><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA()</a></code>] -&gt;
[<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors()</a></code> -&gt;
<code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code>]</li>
<li>knn graph: [<code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code>]</li>
</ul>
<p><code><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA()</a></code> is sometimes run even on dimension reduction
objects (within scoring functions) because some scores require a
variance associated with each dimension.</p>
<p>Let’s process all the outputs. Here, we will go through all the steps
for a more exhaustive demonstration. However, it is to be noted that
skipping the final step <code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code> makes the
neighbour graph computation step (<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors()</a></code>)
unnecessary. In such a case however, one will forgo two scoring metrics,
namely ARI and NMI.</p>
<p>Here, we will use <code><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn()</a></code> between
<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors()</a></code> and <code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code>
because we set <code>return.neighbor = TRUE</code> in
<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors()</a></code>. This is useful to keep the distances
between cells in the <code>KNN</code> graph rather than what
<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors()</a></code> does by default, which is converting the
<code>KNN</code> graph to an adjacency matrix with 0/1s and to a
<code>SNN</code> network with values bounded between 0 and 1. This not
compulsory, but this is used to stay in line with BBKNN’s output. To
prevent the community detection algorithm to output a high fraction of
singletons, we “<em>symmetrize</em>” the matrix which makes the graph
“undirected”.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># corrected counts outputs</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"scanorama.reconstructed"</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seu</span>, npcs <span class="op">=</span> <span class="fl">50L</span>, reduction.name <span class="op">=</span> <span class="st">"pca.scanorama_counts"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca.scanorama_counts"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     graph.name <span class="op">=</span> <span class="st">"knn.scanorama_counts"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.scanorama_counts"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.scanorama_counts_symmetric"</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"scanorama_counts_{cell.var}_{metric}"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           optimisation.metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nmi"</span>, <span class="st">"ari"</span><span class="op">)</span><span class="op">)</span> <span class="co"># default, compute both</span></span>
<span></span>
<span></span>
<span><span class="co"># dimension reduction outputs</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT"</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, k.param <span class="op">=</span> <span class="fl">20L</span>,</span>
<span>                     return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>, graph.name <span class="op">=</span> <span class="st">"knn.unintegrated"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.unintegrated"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.unintegrated_symmetric"</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"unintegrated_{cell.var}_{metric}"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"integrated.scanorama"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>                     return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>, graph.name <span class="op">=</span> <span class="st">"knn.scanorama_reduction"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.scanorama_reduction"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.scanorama_reduction_symmetric"</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"scanorama_reduction_{cell.var}_{metric}"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>                     return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>, graph.name <span class="op">=</span> <span class="st">"knn.harmony"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.harmony"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.harmony_symmetric"</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"harmony_{cell.var}_{metric}"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"cca.integrated"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>                     return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>, graph.name <span class="op">=</span> <span class="st">"knn.cca"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.cca"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.cca_symmetric"</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"cca_{cell.var}_{metric}"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"rpca.integrated"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>,</span>
<span>                     return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>, graph.name <span class="op">=</span> <span class="st">"knn.rpca"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.rpca"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.rpca_symmetric"</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"rpca_{cell.var}_{metric}"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="st">"integrated.scANVI"</span>,</span>
<span>                     return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>, graph.name <span class="op">=</span> <span class="st">"knn.scanvi"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.scanvi"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"knn.scanvi_symmetric"</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"scanvi_{cell.var}_{metric}"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># graph outputs</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"bbknn_ridge.residuals_distances"</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">seu</span>, graph.name <span class="op">=</span> <span class="st">"bbknn_ridge.residuals_distances_symmetric"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>, cluster.name <span class="op">=</span> <span class="st">"bbknn_{cell.var}_{metric}"</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p><em>Note</em>: Instead of sticking with the default
<code>FindNeighbors(return.neighbors = FALSE)</code> at <a href="#standard-worflow">the beginning</a>, we could have switched it to
<code>TRUE</code> right away, process the <code>KNN</code> graph with
<code><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn()</a></code> and use it for subsequent steps (umap,
clustering, etc.)</p>
</blockquote>
<p><code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code> adds metadata columns with the
clustering results that maximized a metric (NMI or ARI) to the Seurat
object:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/paged_table.html" class="external-link">paged_table</a></span><span class="op">(</span><span class="va">seu</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"CellType_[arinm]{3}$"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seu</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["scanorama_counts_CellType_nmi"],"name":[1],"type":["fct"],"align":["left"]},{"label":["scanorama_counts_CellType_ari"],"name":[2],"type":["fct"],"align":["left"]},{"label":["unintegrated_CellType_nmi"],"name":[3],"type":["fct"],"align":["left"]},{"label":["unintegrated_CellType_ari"],"name":[4],"type":["fct"],"align":["left"]},{"label":["scanorama_reduction_CellType_nmi"],"name":[5],"type":["fct"],"align":["left"]},{"label":["scanorama_reduction_CellType_ari"],"name":[6],"type":["fct"],"align":["left"]},{"label":["harmony_CellType_nmi"],"name":[7],"type":["fct"],"align":["left"]},{"label":["harmony_CellType_ari"],"name":[8],"type":["fct"],"align":["left"]},{"label":["cca_CellType_nmi"],"name":[9],"type":["fct"],"align":["left"]},{"label":["cca_CellType_ari"],"name":[10],"type":["fct"],"align":["left"]},{"label":["rpca_CellType_nmi"],"name":[11],"type":["fct"],"align":["left"]},{"label":["rpca_CellType_ari"],"name":[12],"type":["fct"],"align":["left"]},{"label":["scanvi_CellType_nmi"],"name":[13],"type":["fct"],"align":["left"]},{"label":["scanvi_CellType_ari"],"name":[14],"type":["fct"],"align":["left"]},{"label":["bbknn_CellType_nmi"],"name":[15],"type":["fct"],"align":["left"]},{"label":["bbknn_CellType_ari"],"name":[16],"type":["fct"],"align":["left"]}],"data":[{"1":"3","2":"3","3":"1","4":"1","5":"1","6":"1","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_108"},{"1":"1","2":"1","3":"1","4":"1","5":"1","6":"1","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_115"},{"1":"3","2":"3","3":"1","4":"1","5":"1","6":"1","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_133"},{"1":"1","2":"1","3":"1","4":"1","5":"2","6":"2","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_142"},{"1":"3","2":"3","3":"1","4":"1","5":"1","6":"1","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_143"},{"1":"4","2":"4","3":"1","4":"1","5":"2","6":"2","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_144"},{"1":"1","2":"1","3":"1","4":"1","5":"1","6":"1","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_146"},{"1":"1","2":"1","3":"1","4":"1","5":"2","6":"1","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_148"},{"1":"1","2":"1","3":"1","4":"1","5":"1","6":"1","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_152"},{"1":"1","2":"1","3":"1","4":"1","5":"2","6":"1","7":"1","8":"1","9":"0","10":"0","11":"1","12":"1","13":"2","14":"3","15":"1","16":"1","_rn_":"pbmc1_SM2_Cell_153"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="section level3">
<h3 id="add-scores-to-the-seurat-object">Add scores to the Seurat object<a class="anchor" aria-label="anchor" href="#add-scores-to-the-seurat-object"></a>
</h3>
<p>Now that we have computed objects ready to be scored, we will proceed
with assessing each integration output. For this step, one can either
use <code>Score[score_name]()</code> and save scores in separate
variables or use <code>AddScore[score_name]()</code> to directly store
scores within the Seurat object. The latter is far more convenient and
allows to compare scores graphically. This is the strategy we are going
to adopt here.</p>
<p>Last but not least, a cell-type variable is used in most scores,
hence it is highly recommended to have an estimate of each cell’s type
(or produce such an estimate) stored in the Seurat object as a column in
the metadata.</p>
<blockquote>
<p>Note that if one doesn’t have a variable with cell-labels (used as
ground truth in multiple scores) he will have to produce it or do
without several scores. Alternatively, one can use automatic cell
annotation algorithms (e.g. the <a href="https://github.com/satijalab/azimuth" class="external-link">Azimuth package</a>). One
can also have multiple cell label variables (e.g. Azimuth typically
returns as many cell label variables as levels of annotations contained
in the reference). Scores requiring cell type annotations always accept
multiple column names.</p>
</blockquote>
<blockquote>
<p>There is a risk of confusion between cell annotations when using
automatic cell annotation tools. Furthermore, in the case of using
Azimuth to annotate cells, there is a specific risk of biasing results
by favouring RPCA integration because Azimuth uses RPCA to integrate the
query dataset onto the reference.</p>
</blockquote>
<p>First of all, let’s organise outputs in lists:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reductions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  unintegrated <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>  scanorama_counts <span class="op">=</span> <span class="st">"pca.scanorama_counts"</span>,</span>
<span>  scanorama_reduction <span class="op">=</span> <span class="st">"integrated.scanorama"</span>,</span>
<span>  harmony <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>  cca <span class="op">=</span> <span class="st">"cca.integrated"</span>,</span>
<span>  rpca <span class="op">=</span> <span class="st">"rpca.integrated"</span>,</span>
<span>  scanvi <span class="op">=</span> <span class="st">"integrated.scANVI"</span>,</span>
<span>  bbknn <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">graphs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  unintegrated <span class="op">=</span> <span class="st">"knn.unintegrated_symmetric"</span>,</span>
<span>  scanorama_counts <span class="op">=</span> <span class="st">"knn.scanorama_counts_symmetric"</span>,</span>
<span>  scanorama_reduction <span class="op">=</span> <span class="st">"knn.scanorama_reduction_symmetric"</span>,</span>
<span>  harmony <span class="op">=</span> <span class="st">"knn.harmony_symmetric"</span>,</span>
<span>  cca <span class="op">=</span> <span class="st">"knn.cca_symmetric"</span>,</span>
<span>  rpca <span class="op">=</span> <span class="st">"knn.rpca_symmetric"</span>,</span>
<span>  scanvi <span class="op">=</span> <span class="st">"knn.scanvi_symmetric"</span>,</span>
<span>  bbknn <span class="op">=</span> <span class="st">"bbknn_ridge.residuals_distances_symmetric"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">integrations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">reductions</span><span class="op">)</span></span></code></pre></div>
<p>Let’s finalise our preparations: :</p>
<ul>
<li>Make sure you have the <a href="https://github.com/theislab/kBET" class="external-link">kBET package</a> installed to
run <code><a href="../reference/score-kbet.html">ScoreKBET()</a></code>
</li>
<li>To benefit from a faster implementation of the lisi algorithm,
install the <a href="https://github.com/immunogenomics/LISI" class="external-link">lisi
package</a>
</li>
<li>To benefit from a faster implementation of the euclidean distance
computation, install the <a href="https://doi.com/10.32614/CRAN.package.distances" class="external-link">distances
package</a>
</li>
<li>To run <code><a href="../reference/score-cc.html">ScoreRegressPC.CellCycle()</a></code> (score of cell cycle
conservation), run <code><a href="../reference/CellCycleScoringPerBatch.html">CellCycleScoringPerBatch()</a></code> beforehand
(next chunk)</li>
</ul>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CellCycleScoringPerBatch.html">CellCycleScoringPerBatch</a></span><span class="op">(</span><span class="va">seu</span>, batch.var <span class="op">=</span> <span class="va">batch.var</span>,</span>
<span>                                s.features <span class="op">=</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span>,</span>
<span>                                g2m.features <span class="op">=</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span><span class="op">)</span></span></code></pre></div>
<p>Let’s now loop through the integration outputs:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">integ</span> <span class="kw">in</span> <span class="va">integrations</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">reduc</span> <span class="op">&lt;-</span> <span class="va">reductions</span><span class="op">[[</span><span class="va">integ</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">graph</span> <span class="op">&lt;-</span> <span class="va">graphs</span><span class="op">[[</span><span class="va">integ</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">clust.var.ari</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">integ</span>, <span class="va">cell.var</span>, <span class="st">"ari"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span> <span class="co"># produced by `FindOptimalClusters()`</span></span>
<span>  <span class="va">clust.var.nmi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">integ</span>, <span class="va">cell.var</span>, <span class="st">"nmi"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span> <span class="co"># produced by `FindOptimalClusters()`</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">reduc</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>   <span class="co"># all TRUE except bbknn</span></span>
<span>    <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-asw.html">AddScoreASW</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                       what <span class="op">=</span> <span class="va">reduc</span>, dist.package <span class="op">=</span> <span class="st">"distances"</span><span class="op">)</span></span>
<span>    <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-asw.html">AddScoreASWBatch</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>, batch.var <span class="op">=</span> <span class="va">batch.var</span>,</span>
<span>                            cell.var <span class="op">=</span> <span class="va">cell.var</span>, what <span class="op">=</span> <span class="va">reduc</span>,</span>
<span>                            dist.package <span class="op">=</span> <span class="st">"distances"</span><span class="op">)</span></span>
<span>    <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-densityPC.html">AddScoreDensityPC</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>, batch.var <span class="op">=</span> <span class="va">batch.var</span>,</span>
<span>                             reduction <span class="op">=</span> <span class="va">reduc</span><span class="op">)</span></span>
<span>    <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-regressPC.html">AddScoreRegressPC</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>, batch.var <span class="op">=</span> <span class="va">batch.var</span>,</span>
<span>                             reduction <span class="op">=</span> <span class="va">reduc</span><span class="op">)</span></span>
<span>    <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-cc.html">AddScoreRegressPC.CellCycle</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>,</span>
<span>                                       batch.var <span class="op">=</span> <span class="va">batch.var</span>, what <span class="op">=</span> <span class="va">reduc</span>,</span>
<span>                                       compute.cc <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># because CellCycleScoringPerBatch was ran before</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-ari.html">AddScoreARI</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                     clust.var <span class="op">=</span> <span class="va">clust.var.ari</span><span class="op">)</span></span>
<span>  <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-nmi.html">AddScoreNMI</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                     clust.var <span class="op">=</span> <span class="va">clust.var.nmi</span><span class="op">)</span></span>
<span>  <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-connectivity.html">AddScoreConnectivity</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>, graph.name <span class="op">=</span> <span class="va">graph</span>,</span>
<span>                              cell.var <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span></span>
<span>  <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-kbet.html">AddScoreKBET</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>, batch.var <span class="op">=</span> <span class="va">batch.var</span>,</span>
<span>                      cell.var <span class="op">=</span> <span class="va">cell.var</span>, what <span class="op">=</span> <span class="va">reduc</span> <span class="op"><a href="https://rdrr.io/r/base/Control.html" class="external-link">%||%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"_symmetric$"</span>, <span class="st">""</span>, <span class="va">graph</span><span class="op">)</span>,</span>
<span>                      graph.type <span class="op">=</span> <span class="st">"distances"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-lisi.html">AddScoreLISI</a></span><span class="op">(</span><span class="va">seu</span>, integration <span class="op">=</span> <span class="va">integ</span>, batch.var <span class="op">=</span> <span class="va">batch.var</span>,</span>
<span>                      cell.var <span class="op">=</span> <span class="va">cell.var</span>, reduction <span class="op">=</span> <span class="va">reduc</span>,</span>
<span>                      graph.name <span class="op">=</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">reduc</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"_symmetric$"</span>, <span class="st">""</span>, <span class="va">graph</span><span class="op">)</span> <span class="kw">else</span> <span class="cn">NULL</span>,</span>
<span>                      graph.type <span class="op">=</span> <span class="st">"distances"</span>, ncores <span class="op">=</span> <span class="va">ncores</span><span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now that we have computed multiple scores, we can look at them using
<code><a href="../reference/get-score.html">IntegrationScores()</a></code>:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/paged_table.html" class="external-link">paged_table</a></span><span class="op">(</span><span class="fu"><a href="../reference/get-score.html">IntegrationScores</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Integration"],"name":[1],"type":["chr"],"align":["left"]},{"label":["PCA.regression"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["PCA.density"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["cell.cycle.conservation"],"name":[4],"type":["list"],"align":["right"]},{"label":["ASW"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["ASW.batch"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["NMI"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["ARI"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["Graph.connectivity"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["ASW_CellType"],"name":[10],"type":["dbl"],"align":["right"]},{"label":["ASW.batch_CellType"],"name":[11],"type":["dbl"],"align":["right"]},{"label":["ARI_CellType"],"name":[12],"type":["dbl"],"align":["right"]},{"label":["NMI_CellType"],"name":[13],"type":["dbl"],"align":["right"]},{"label":["Graph.connectivity_CellType"],"name":[14],"type":["dbl"],"align":["right"]},{"label":["kBET_CellType"],"name":[15],"type":["dbl"],"align":["right"]},{"label":["iLISI_Method"],"name":[16],"type":["numrc_ls"],"align":["right"]},{"label":["cLISI_CellType"],"name":[17],"type":["numrc_ls"],"align":["right"]}],"data":[{"1":"Unintegrated","2":"9.938453e-33","3":"0.42889628","4":"<tibble[,2]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"0.5543013","11":"0.7655136","12":"0.7586378","13":"0.7887182","14":"0.9534725","15":"0.5998129","16":"1.058542","17":"1.018809"},{"1":"CCA","2":"2.447451e-03","3":"0.48887307","4":"<tibble[,2]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"0.5066540","11":"0.8250420","12":"0.5919543","13":"0.6871637","14":"0.9658361","15":"0.4807587","16":"1.595616","17":"1.418842"},{"1":"RPCA","2":"1.246080e-01","3":"0.41347618","4":"<tibble[,2]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"0.5787869","11":"0.8487809","12":"0.7446937","13":"0.7947530","14":"0.9424623","15":"0.5216103","16":"1.232795","17":"1.006020"},{"1":"Harmony","2":"4.617056e-02","3":"0.45488749","4":"<tibble[,2]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"0.5977252","11":"0.8472633","12":"0.6714181","13":"0.7369587","14":"0.9389699","15":"0.4194419","16":"1.395941","17":"1.021301"},{"1":"MNN","2":"NA","3":"NA","4":"<lgl [1]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"NA","11":"NA","12":"NA","13":"NA","14":"NA","15":"NA","16":"NA","17":"NA"},{"1":"FastMNN","2":"NA","3":"NA","4":"<lgl [1]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"NA","11":"NA","12":"NA","13":"NA","14":"NA","15":"NA","16":"NA","17":"NA"},{"1":"BBKNN","2":"NA","3":"NA","4":"<lgl [1]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"NA","11":"NA","12":"0.5608836","13":"0.6045187","14":"0.8759149","15":"0.5314811","16":"1.984795","17":"1.295424"},{"1":"ComBat","2":"NA","3":"NA","4":"<lgl [1]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"NA","11":"NA","12":"NA","13":"NA","14":"NA","15":"NA","16":"NA","17":"NA"},{"1":"Scanorama_reduction","2":"1.606113e-01","3":"0.34479126","4":"<tibble[,2]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"0.5116296","11":"0.7832223","12":"0.6525726","13":"0.5908032","14":"0.9225986","15":"0.7542150","16":"1.067041","17":"1.188885"},{"1":"Scanorama_counts","2":"1.299169e-01","3":"0.32499565","4":"<tibble[,2]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"0.5275665","11":"0.6732213","12":"0.6207683","13":"0.6450650","14":"0.9048762","15":"0.7351793","16":"1.049071","17":"1.098436"},{"1":"scVI","2":"NA","3":"NA","4":"<lgl [1]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"NA","11":"NA","12":"NA","13":"NA","14":"NA","15":"NA","16":"NA","17":"NA"},{"1":"scANVI","2":"8.550245e-01","3":"0.01271029","4":"<tibble[,2]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"0.2776570","11":"0.2049291","12":"0.6505520","13":"0.6305290","14":"0.7346842","15":"0.9084372","16":"1.000000","17":"1.117207"},{"1":"trVAE","2":"NA","3":"NA","4":"<lgl [1]>","5":"NA","6":"NA","7":"NA","8":"NA","9":"NA","10":"NA","11":"NA","12":"NA","13":"NA","14":"NA","15":"NA","16":"NA","17":"NA"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div class="section level3">
<h3 id="compare-the-integrations">Compare the integrations<a class="anchor" aria-label="anchor" href="#compare-the-integrations"></a>
</h3>
<p>Note that those scores are raw. Let’s scale them (and make them
comparable):</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScaleScores.html">ScaleScores</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<p>Once again, we can print them:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">rmarkdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/paged_table.html" class="external-link">paged_table</a></span><span class="op">(</span><span class="fu"><a href="../reference/get-score.html">IntegrationScores</a></span><span class="op">(</span><span class="va">seu</span>, scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["Integration"],"name":[1],"type":["chr"],"align":["left"]},{"label":["PCA.regression"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["PCA.density"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["cell.cycle.conservation"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["ASW_CellType"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["ASW.batch_CellType"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["ARI_CellType"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["NMI_CellType"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["Graph.connectivity_CellType"],"name":[9],"type":["dbl"],"align":["right"]},{"label":["kBET_CellType"],"name":[10],"type":["dbl"],"align":["right"]},{"label":["iLISI_Method"],"name":[11],"type":["numrc_ls"],"align":["right"]},{"label":["cLISI_CellType"],"name":[12],"type":["numrc_ls"],"align":["right"]},{"label":["Bio.conservation"],"name":[13],"type":["dbl"],"align":["right"]},{"label":["Batch.correction"],"name":[14],"type":["dbl"],"align":["right"]},{"label":["Overall.score"],"name":[15],"type":["dbl"],"align":["right"]}],"data":[{"1":"Unintegrated","2":"0","3":"0.00000000","4":"1.0000000","5":"0.5543013","6":"0.7655136","7":"0.7586378","8":"0.7887182","9":"0.9534725","10":"0.40018708","11":"0.02927093","12":"0.9976489","13":"0.8198612","14":"0.3580740","15":"0.6351463"},{"1":"CCA","2":"0","3":"0.12268377","4":"0.7545196","5":"0.5066540","6":"0.8250420","7":"0.5919543","8":"0.6871637","9":"0.9658361","10":"0.51924134","11":"0.29780811","12":"0.9476447","13":"0.6975873","14":"0.4551019","15":"0.6005931"},{"1":"RPCA","2":"0","3":"0.00000000","4":"0.9295404","5":"0.5787869","6":"0.8487809","7":"0.7446937","8":"0.7947530","9":"0.9424623","10":"0.47838966","11":"0.11639749","12":"0.9992476","13":"0.8094043","14":"0.3976717","15":"0.6447113"},{"1":"Harmony","2":"0","3":"0.05713768","4":"0.8817346","5":"0.5977252","6":"0.8472633","7":"0.6714181","8":"0.7369587","9":"0.9389699","10":"0.58055814","11":"0.19797066","12":"0.9973374","13":"0.7770348","14":"0.4369833","15":"0.6410142"},{"1":"BBKNN","2":"NA","3":"NA","4":"NA","5":"NA","6":"NA","7":"0.5608836","8":"0.6045187","9":"0.8759149","10":"0.46851886","11":"0.49239764","12":"0.9630720","13":"0.7094914","14":"0.6122771","15":"0.6706057"},{"1":"Scanorama_reduction","2":"0","3":"0.00000000","4":"0.7731061","5":"0.5116296","6":"0.7832223","7":"0.6525726","8":"0.5908032","9":"0.9225986","10":"0.24578501","11":"0.03352049","12":"0.9763894","13":"0.7009002","14":"0.3308544","15":"0.5528819"},{"1":"Scanorama_counts","2":"0","3":"0.00000000","4":"0.6434536","5":"0.5275665","6":"0.6732213","7":"0.6207683","8":"0.6450650","9":"0.9048762","10":"0.26482067","11":"0.02453554","12":"0.9876955","13":"0.6849098","14":"0.3112423","15":"0.5354428"},{"1":"scANVI","2":"0","3":"0.00000000","4":"0.4274816","5":"0.2776570","6":"0.2049291","7":"0.6505520","8":"0.6305290","9":"0.7346842","10":"0.09156279","11":"0.00000000","12":"0.9853491","13":"0.5943137","14":"0.1718627","15":"0.4253333"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>To readily compare the integrations, let’s plot the scores:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotScores.html">PlotScores</a></span><span class="op">(</span><span class="va">seu</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-21-1.png" alt="Dot plot to compare performance of integrations based on scores" width="864"></p>
<p>One might notice a difference in scale for some of the scores, if
comparing the plot with the table just before. This is the case for the
PCA density score for instance.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="../reference/get-score.html">IntegrationScores</a></span><span class="op">(</span><span class="va">seu</span>, scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">$</span><span class="va">PCA.density</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.00000000 0.00000000 0.00000000 0.00000000 0.00000000 0.05713768 0.12268377</span></span></code></pre>
<p>Indeed, <code><a href="../reference/PlotScores.html">PlotScores()</a></code> rescales the scores using min-max
normalisation by default (<code>rescale = TRUE</code>). One might chose
to disable it:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotScores.html">PlotScores</a></span><span class="op">(</span><span class="va">seu</span>, rescale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-23-1.png" alt="Dot plot of scores without min-max rescaling" width="864"></p>
<p>We notice that PCA based methods output very low scores in this case.
Since they cannot be computed on knn graphs, scores are biased in favour
of BBKNN. You can exclude some scores (and recompute overall scores on
the fly)</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotScores.html">PlotScores</a></span><span class="op">(</span><span class="va">seu</span>, rescale <span class="op">=</span> <span class="cn">FALSE</span>, exclude.score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pca.density"</span>, <span class="st">"pca.regression"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-24-1.png" alt="Dot plot of scores without min-max rescaling, excluding non-informative scores" width="816"></p>
<p>You can chose a different type of plot (<code>radar</code> or
<code>lollipop</code>):</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/PlotScores.html">PlotScores</a></span><span class="op">(</span><span class="va">seu</span>, plot.type <span class="op">=</span> <span class="st">"radar"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># reduce overlap between axis names and figures </span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span>, panel.spacing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>        plot.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span>r <span class="op">=</span> <span class="fl">3</span>, l <span class="op">=</span> <span class="fl">3</span>, unit <span class="op">=</span> <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 5 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_point()`).</span></span></code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-25-1.png" alt="Radar plot of scores" width="1152"></p>
<p>In this last plot, we also exclude the non-integrated dataset. Since
the <code>rescale</code> argument is true by default, the scores are
rescaled without the excluded integration’s scores.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="fu"><a href="../reference/PlotScores.html">PlotScores</a></span><span class="op">(</span><span class="va">seu</span>, plot.type <span class="op">=</span> <span class="st">"lollipop"</span>,</span>
<span>             exclude.integration <span class="op">=</span> <span class="st">"unintegrated"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 5 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_point()`).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 5 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_errorbarh()`).</span></span></code></pre>
<p><img src="introduction_files/figure-html/unnamed-chunk-26-1.png" alt="Lollipop plot of scores without the unintegrated case" width="864"></p>
<p>We want to compare the UMAP embeddings. For this, we first compute
the dimension reductions:</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">for</span> <span class="op">(</span><span class="va">integ</span> <span class="kw">in</span> <span class="va">integrations</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">reduc</span> <span class="op">&lt;-</span> <span class="va">reductions</span><span class="op">[[</span><span class="va">integ</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">reduc</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span> <span class="co"># all except BBKNN</span></span>
<span>    <span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seu</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">seu</span><span class="op">[[</span><span class="va">reduc</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, reduction <span class="op">=</span> <span class="va">reduc</span>,</span>
<span>                   reduction.name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">integ</span>, <span class="st">".umap"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>As BBKNN’s output is a graph, we need to use the umap Python
package:</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/conda-tools.html" class="external-link">condaenv_exists</a></span><span class="op">(</span><span class="st">'umap_0.5.4'</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/conda-tools.html" class="external-link">conda_create</a></span><span class="op">(</span><span class="st">'umap_0.5.4'</span>, packages <span class="op">=</span> <span class="st">'umap=0.5.4'</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span><span class="op">)</span></span>
<span><span class="va">seu</span> <span class="op"><a href="https://future.futureverse.org/reference/future.html" class="external-link">%&lt;-%</a></span> <span class="op">{</span></span>
<span>  <span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/use_python.html" class="external-link">use_condaenv</a></span><span class="op">(</span><span class="st">'umap_0.5.4'</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">seu</span>, graph <span class="op">=</span> <span class="st">"bbknn_ridge.residuals_connectivities"</span>, umap.method <span class="op">=</span> <span class="st">"umap-learn"</span>,</span>
<span>          n.epochs <span class="op">=</span> <span class="fl">200L</span>, reduction.name <span class="op">=</span> <span class="st">"bbknn.umap"</span><span class="op">)</span> <span class="op">}</span></span>
<span><span class="va">seu</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">sequential</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">integrations</span>, <span class="kw">function</span><span class="op">(</span><span class="va">integ</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">integ</span>, <span class="st">".umap"</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="va">batch.var</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">integ</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-29-1.png" alt="UMAPs of unintegrated and integrated pbmcsca data with cells colored by sequencing method (i.e. batch)" width="960"></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">plot_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">integrations</span>, <span class="kw">function</span><span class="op">(</span><span class="va">integ</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">seu</span>, reduction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">integ</span>, <span class="st">".umap"</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="va">integ</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span>, simplify <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plot_list</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-30-1.png" alt="UMAPs of unintegrated and integrated pbmcsca data with cells colored by cell type label" width="960"></p>
<p>There are several observations to be made here, that require further
explanations:</p>
<ul>
<li>some cells seem to be assigned a wrong label in
<code>CellType</code>, highlighting the importance of having cell
annotations of sufficient quality to be considered suitable as ground
truth (which is actually not the case here)</li>
<li>All scaled PCA regression scores are set to zero. This is because
the unintegrated dataset has the lowest raw PCA regression score. It is
very likely the consequence of the <code>SCT</code> normalisation, which
is much more efficient at masking batch-specific differences than the
classical <code>LogNorm</code>. Thus, the inter-batch differences are
not driving the principal components.</li>
<li>
<code><a href="../reference/ScaleScores.html">ScaleScores()</a></code> produce scores that can have different
scales (as long as <code>rescale = FALSE</code>) . Thus, min-max
rescaling is used by default in <code><a href="../reference/PlotScores.html">PlotScores()</a></code>, to balance
each score’s contribution to the overall scores. This is especially
suited for comparing a large number of integrations. However, it has
some drawbacks: it can heavily distort the scale of scores when their
maximum or minimum are far from 1 or 0 respectively (e.g. all cLISI
scores are above 0.9). Hence, the <strong>final decision</strong> on
whether to use min-max rescaling <strong>is left to the user’s
discretion</strong>.</li>
</ul>
<details><summary><strong>Session info</strong>
</summary><pre><code><span>  <span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span>  <span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span>  <span class="co">## Running under: Ubuntu 24.04.1 LTS</span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## Matrix products: default</span></span>
<span>  <span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span>  <span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## locale:</span></span>
<span>  <span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span>  <span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span>  <span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span>  <span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## time zone: UTC</span></span>
<span>  <span class="co">## tzcode source: system (glibc)</span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## attached base packages:</span></span>
<span>  <span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## other attached packages:</span></span>
<span>  <span class="co">## [1] future_1.34.0         ggplot2_3.5.1         SeuratIntegrate_0.4.0</span></span>
<span>  <span class="co">## [4] Seurat_5.2.1          SeuratObject_5.0.2    sp_2.1-4             </span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## loaded via a namespace (and not attached):</span></span>
<span>  <span class="co">##   [1] RcppAnnoy_0.0.22            splines_4.4.2              </span></span>
<span>  <span class="co">##   [3] later_1.4.1                 batchelor_1.22.0           </span></span>
<span>  <span class="co">##   [5] tibble_3.2.1                polyclip_1.10-7            </span></span>
<span>  <span class="co">##   [7] XML_3.99-0.18               fastDummies_1.7.5          </span></span>
<span>  <span class="co">##   [9] lifecycle_1.0.4             edgeR_4.4.1                </span></span>
<span>  <span class="co">##  [11] globals_0.16.3              lattice_0.22-6             </span></span>
<span>  <span class="co">##  [13] MASS_7.3-61                 backports_1.5.0            </span></span>
<span>  <span class="co">##  [15] magrittr_2.0.3              limma_3.62.2               </span></span>
<span>  <span class="co">##  [17] plotly_4.10.4               sass_0.4.9                 </span></span>
<span>  <span class="co">##  [19] rmarkdown_2.29              jquerylib_0.1.4            </span></span>
<span>  <span class="co">##  [21] yaml_2.3.10                 httpuv_1.6.15              </span></span>
<span>  <span class="co">##  [23] glmGamPoi_1.18.0            sctransform_0.4.1          </span></span>
<span>  <span class="co">##  [25] spam_2.11-1                 spatstat.sparse_3.1-0      </span></span>
<span>  <span class="co">##  [27] reticulate_1.40.0           DBI_1.2.3                  </span></span>
<span>  <span class="co">##  [29] cowplot_1.1.3               pbapply_1.7-2              </span></span>
<span>  <span class="co">##  [31] RColorBrewer_1.1-3          ResidualMatrix_1.16.0      </span></span>
<span>  <span class="co">##  [33] abind_1.4-8                 zlibbioc_1.52.0            </span></span>
<span>  <span class="co">##  [35] Rtsne_0.17                  GenomicRanges_1.58.0       </span></span>
<span>  <span class="co">##  [37] purrr_1.0.2                 BiocGenerics_0.52.0        </span></span>
<span>  <span class="co">##  [39] tweenr_2.0.3                rappdirs_0.3.3             </span></span>
<span>  <span class="co">##  [41] sva_3.54.0                  GenomeInfoDbData_1.2.13    </span></span>
<span>  <span class="co">##  [43] IRanges_2.40.1              S4Vectors_0.44.0           </span></span>
<span>  <span class="co">##  [45] ggrepel_0.9.6               irlba_2.3.5.1              </span></span>
<span>  <span class="co">##  [47] listenv_0.9.1               spatstat.utils_3.1-2       </span></span>
<span>  <span class="co">##  [49] genefilter_1.88.0           pbmcsca.SeuratData_3.0.0   </span></span>
<span>  <span class="co">##  [51] goftest_1.2-3               RSpectra_0.16-2            </span></span>
<span>  <span class="co">##  [53] annotate_1.84.0             spatstat.random_3.3-2      </span></span>
<span>  <span class="co">##  [55] fitdistrplus_1.2-2          parallelly_1.41.0          </span></span>
<span>  <span class="co">##  [57] pkgdown_2.1.1               DelayedMatrixStats_1.28.1  </span></span>
<span>  <span class="co">##  [59] codetools_0.2-20            DelayedArray_0.32.0        </span></span>
<span>  <span class="co">##  [61] ggforce_0.4.2               scuttle_1.16.0             </span></span>
<span>  <span class="co">##  [63] tidyselect_1.2.1            UCSC.utils_1.2.0           </span></span>
<span>  <span class="co">##  [65] farver_2.1.2                ScaledMatrix_1.14.0        </span></span>
<span>  <span class="co">##  [67] matrixStats_1.5.0           stats4_4.4.2               </span></span>
<span>  <span class="co">##  [69] spatstat.explore_3.3-4      jsonlite_1.8.9             </span></span>
<span>  <span class="co">##  [71] BiocNeighbors_2.0.1         progressr_0.15.1           </span></span>
<span>  <span class="co">##  [73] ggridges_0.5.6              survival_3.7-0             </span></span>
<span>  <span class="co">##  [75] systemfonts_1.2.1           tools_4.4.2                </span></span>
<span>  <span class="co">##  [77] ragg_1.3.3                  ica_1.0-3                  </span></span>
<span>  <span class="co">##  [79] Rcpp_1.0.14                 glue_1.8.0                 </span></span>
<span>  <span class="co">##  [81] gridExtra_2.3               SparseArray_1.6.1          </span></span>
<span>  <span class="co">##  [83] mgcv_1.9-1                  xfun_0.50                  </span></span>
<span>  <span class="co">##  [85] MatrixGenerics_1.18.1       GenomeInfoDb_1.42.1        </span></span>
<span>  <span class="co">##  [87] dplyr_1.1.4                 withr_3.0.2                </span></span>
<span>  <span class="co">##  [89] fastmap_1.2.0               digest_0.6.37              </span></span>
<span>  <span class="co">##  [91] rsvd_1.0.5                  R6_2.5.1                   </span></span>
<span>  <span class="co">##  [93] mime_0.12                   textshaping_1.0.0          </span></span>
<span>  <span class="co">##  [95] colorspace_2.1-1            scattermore_1.2            </span></span>
<span>  <span class="co">##  [97] tensor_1.5                  kBET_0.99.6                </span></span>
<span>  <span class="co">##  [99] RSQLite_2.3.9               spatstat.data_3.1-4        </span></span>
<span>  <span class="co">## [101] RhpcBLASctl_0.23-42         tidyr_1.3.1                </span></span>
<span>  <span class="co">## [103] generics_0.1.3              data.table_1.16.4          </span></span>
<span>  <span class="co">## [105] FNN_1.1.4.1                 httr_1.4.7                 </span></span>
<span>  <span class="co">## [107] htmlwidgets_1.6.4           S4Arrays_1.6.0             </span></span>
<span>  <span class="co">## [109] uwot_0.2.2                  pkgconfig_2.0.3            </span></span>
<span>  <span class="co">## [111] gtable_0.3.6                blob_1.2.4                 </span></span>
<span>  <span class="co">## [113] lmtest_0.9-40               SingleCellExperiment_1.28.1</span></span>
<span>  <span class="co">## [115] XVector_0.46.0              htmltools_0.5.8.1          </span></span>
<span>  <span class="co">## [117] dotCall64_1.2               scales_1.3.0               </span></span>
<span>  <span class="co">## [119] Biobase_2.66.0              SeuratData_0.2.2.9001      </span></span>
<span>  <span class="co">## [121] png_0.1-8                   lisi_1.0                   </span></span>
<span>  <span class="co">## [123] harmony_1.2.3               spatstat.univar_3.1-1      </span></span>
<span>  <span class="co">## [125] knitr_1.49                  reshape2_1.4.4             </span></span>
<span>  <span class="co">## [127] nlme_3.1-166                cachem_1.1.0               </span></span>
<span>  <span class="co">## [129] zoo_1.8-12                  stringr_1.5.1              </span></span>
<span>  <span class="co">## [131] KernSmooth_2.23-24          parallel_4.4.2             </span></span>
<span>  <span class="co">## [133] miniUI_0.1.1.1              AnnotationDbi_1.68.0       </span></span>
<span>  <span class="co">## [135] desc_1.4.3                  pillar_1.10.1              </span></span>
<span>  <span class="co">## [137] grid_4.4.2                  vctrs_0.6.5                </span></span>
<span>  <span class="co">## [139] RANN_2.6.2                  promises_1.3.2             </span></span>
<span>  <span class="co">## [141] BiocSingular_1.22.0         distances_0.1.11           </span></span>
<span>  <span class="co">## [143] beachmat_2.22.0             xtable_1.8-4               </span></span>
<span>  <span class="co">## [145] cluster_2.1.6               evaluate_1.0.3             </span></span>
<span>  <span class="co">## [147] locfit_1.5-9.10             cli_3.6.3                  </span></span>
<span>  <span class="co">## [149] compiler_4.4.2              rlang_1.1.5                </span></span>
<span>  <span class="co">## [151] crayon_1.5.3                future.apply_1.11.3        </span></span>
<span>  <span class="co">## [153] labeling_0.4.3              plyr_1.8.9                 </span></span>
<span>  <span class="co">## [155] forcats_1.0.0               fs_1.6.5                   </span></span>
<span>  <span class="co">## [157] stringi_1.8.4               viridisLite_0.4.2          </span></span>
<span>  <span class="co">## [159] deldir_2.0-4                BiocParallel_1.40.0        </span></span>
<span>  <span class="co">## [161] Biostrings_2.74.1           munsell_0.5.1              </span></span>
<span>  <span class="co">## [163] lazyeval_0.2.2              spatstat.geom_3.3-5        </span></span>
<span>  <span class="co">## [165] Matrix_1.7-1                RcppHNSW_0.6.0             </span></span>
<span>  <span class="co">## [167] patchwork_1.3.0             bit64_4.6.0-1              </span></span>
<span>  <span class="co">## [169] sparseMatrixStats_1.18.0    statmod_1.5.0              </span></span>
<span>  <span class="co">## [171] KEGGREST_1.46.0             shiny_1.10.0               </span></span>
<span>  <span class="co">## [173] SummarizedExperiment_1.36.0 ROCR_1.0-11                </span></span>
<span>  <span class="co">## [175] memoise_2.0.1               igraph_2.1.4               </span></span>
<span>  <span class="co">## [177] broom_1.0.7                 bslib_0.8.0                </span></span>
<span>  <span class="co">## [179] bit_4.5.0.1</span></span></code></pre>
</details>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Florian Specque, Domitille Chalopin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
