<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to SeuratIntegrate • SeuratIntegrate</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Introduction to SeuratIntegrate">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">SeuratIntegrate</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/SeuratIntegrate.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/introduction.html">Integration example: pbmcsca data</a></li>
    <li><a class="dropdown-item" href="../articles/setup_and_tips.html">Advice about conda and future</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Aide-mémoires"><span class="fa fa-sticky-note"></span></button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-">
<li><a class="dropdown-item" href="../articles/memo_integration.html">Integration methods</a></li>
    <li><a class="dropdown-item" href="../articles/memo_score.html">Scoring metrics</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cbib/Seurat-Integrate/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="SeuratIntegrate_files/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="SeuratIntegrate_files/pagedtable-1.1/js/pagedtable.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Introduction to SeuratIntegrate</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cbib/Seurat-Integrate/blob/main/vignettes/SeuratIntegrate.Rmd" class="external-link"><code>vignettes/SeuratIntegrate.Rmd</code></a></small>
      <div class="d-none name"><code>SeuratIntegrate.Rmd</code></div>
    </div>

    
    
<style type="text/css">
.darker-code {
  color: #2f2f2f;
}
</style>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>SeuratIntegrate is an R package that aims to extend the pool of
single-cell RNA sequencing (scRNA-seq) integration methods available in
<a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>. Moreover,
SeuratIntegrate provides a set of tools to evaluate the performance of
the integrations produced.</p>
<div class="section level3">
<h3 id="integrations">Integrations<a class="anchor" aria-label="anchor" href="#integrations"></a>
</h3>
<p>SeuratIntegrate provides access to R and Python methods to correct
batch effect:</p>
<table class="table">
<caption>Comprehensive overview of the integration methods provided by
SeuratIntegrate</caption>
<colgroup>
<col width="13%">
<col width="13%">
<col width="24%">
<col width="24%">
<col width="24%">
</colgroup>
<thead><tr class="header">
<th align="left">Method</th>
<th align="center">Type</th>
<th align="center">Implementation</th>
<th align="center">Underlying algorithm</th>
<th align="right">Reference</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ComBat</td>
<td align="center"><img src="img%2Flogo_R.svg" title="R logo" width="40" alt="R logo"></td>
<td align="center">
<a href="https://doi.org/doi:10.18129/B9.bioc.sva" class="external-link">sva</a>
(Bioconductor)</td>
<td align="center">Empirical Bayes adjustment</td>
<td align="right"><a href="https://doi.org/10.1093/biostatistics/kxj037" class="external-link">Johnson et al.,
2007</a></td>
</tr>
<tr class="even">
<td align="left">Harmony</td>
<td align="center"><img src="img%2Flogo_R.svg" title="R logo" width="40" alt="R logo"></td>
<td align="center">
<a href="https://cran.r-project.org/package=harmony" class="external-link">harmony</a>
(CRAN)</td>
<td align="center">Iterative embedding correction</td>
<td align="right"><a href="https://doi.org/10.1038/s41592-019-0619-0" class="external-link">Korsunsky et al.,
2019</a></td>
</tr>
<tr class="odd">
<td align="left">MNN</td>
<td align="center"><img src="img%2Flogo_R.svg" title="R logo" width="40" alt="R logo"></td>
<td align="center">
<a href="https://doi.org/doi:10.18129/B9.bioc.batchelor" class="external-link">batchelor</a>
(Bioconductor)</td>
<td align="center">Mutual Nearest Neighbors</td>
<td align="right"><a href="https://doi.org/10.1038/nbt.4091" class="external-link">Haghverdi
et al., 2018</a></td>
</tr>
<tr class="even">
<td align="left">BBKNN</td>
<td align="center"><img src="img%2Flogo_Python.svg" title="Python logo" width="40" alt="Python logo"></td>
<td align="center">
<a href="https://github.com/Teichlab/bbknn" class="external-link">bbknn</a>
(GitHub)</td>
<td align="center">Batch-balanced nearest neighbors</td>
<td align="right"><a href="https://doi.org/10.1093/bioinformatics/btz625" class="external-link">Polański et al.,
2020</a></td>
</tr>
<tr class="odd">
<td align="left">scVI</td>
<td align="center"><img src="img%2Flogo_Python.svg" title="Python logo" width="40" alt="Python logo"></td>
<td align="center">
<a href="https://github.com/scverse/scvi-tools" class="external-link">scvi-tools</a>
(GitHub)</td>
<td align="center">Variational autoencoder</td>
<td align="right"><a href="https://doi.org/10.1038/s41592-018-0229-2" class="external-link">Lopez et al.,
2018</a></td>
</tr>
<tr class="even">
<td align="left">scANVI</td>
<td align="center"><img src="img%2Flogo_Python.svg" title="Python logo" width="40" alt="Python logo"></td>
<td align="center">
<a href="https://github.com/scverse/scvi-tools" class="external-link">scvi-tools</a>
(GitHub)</td>
<td align="center">Semi-supervised variational autoencoder</td>
<td align="right"><a href="https://doi.org/10.15252/msb.20209620" class="external-link">Xu et
al., 2021</a></td>
</tr>
<tr class="odd">
<td align="left">Scanorama</td>
<td align="center"><img src="img%2Flogo_Python.svg" title="Python logo" width="40" alt="Python logo"></td>
<td align="center">
<a href="https://github.com/brianhie/scanorama" class="external-link">scanorama</a> (GitHub)</td>
<td align="center">Manifold alignment</td>
<td align="right"><a href="https://doi.org/10.1038/s41587-019-0113-3" class="external-link">Hie et al.,
2019</a></td>
</tr>
<tr class="even">
<td align="left">trVAE</td>
<td align="center"><img src="img%2Flogo_Python.svg" title="Python logo" width="40" alt="Python logo"></td>
<td align="center">
<a href="https://github.com/theislab/scarches" class="external-link">scArches</a> (GitHub)</td>
<td align="center">Conditional variational autoencoder</td>
<td align="right"><a href="https://doi.org/10.1093/bioinformatics/btaa800" class="external-link">Lotfollahi et al.,
2020</a></td>
</tr>
</tbody>
</table>
<p>SeuratIntegrate provides a new interface to integrate the layers of
an object: <code><a href="../reference/DoIntegrate.html">DoIntegrate()</a></code>. Moreover, SeuratIntegrate is
compatible with <code>CCA</code> and <code>RPCA</code> (included in <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>) and
<code>FastMNN</code> (from <a href="https://github.com/satijalab/seurat-wrappers" class="external-link">SeuratWrappers</a>)</p>
</div>
<div class="section level3">
<h3 id="scoring">Scoring<a class="anchor" aria-label="anchor" href="#scoring"></a>
</h3>
<p>SeuratIntegrate incorporates 11 scoring metrics: 6 quantify the
degree of batch mixing <img src="img%2Fbatch_correction.svg" title="batch correction" width="30" alt="batch correction"> (<em>batch
correction</em>), while 5 assess the preservation of biological
differences <img src="img%2Fbio_conservation.svg" title="bio-conservation" width="30" alt="bio-conservation"> (<em>bio-conservation</em>) based
on ground truth cell type labels.</p>
<p>Below is a table summarising each score’s input and type:</p>
<table class="table">
<caption>Table summarising the inputs required for each score, and the
type of scores it belongs to.</caption>
<colgroup>
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th align="left">Score name</th>
<th align="center">Require a cell type variable</th>
<th align="center">Require a clustering variable</th>
<th align="center">Input</th>
<th align="right">Score type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Cell cycle regression</td>
<td align="center"></td>
<td align="center"></td>
<td align="center"><img src="img%2Fembedding.svg" title="Dimension reduction" width="60" alt="Dimension reduction"></td>
<td align="right"><img src="img%2Fbio_conservation.svg" title="bio-conservation" width="60" alt="bio-conservation"></td>
</tr>
<tr class="even">
<td align="left">PCA regression</td>
<td align="center"></td>
<td align="center"></td>
<td align="center"><img src="img%2Fembedding.svg" title="Dimension reduction" width="60" alt="Dimension reduction"></td>
<td align="right"><img src="img%2Fbatch_correction.svg" title="batch correction" width="60" alt="batch correction"></td>
</tr>
<tr class="odd">
<td align="left">PCA density</td>
<td align="center"></td>
<td align="center"></td>
<td align="center"><img src="img%2Fembedding.svg" title="Dimension reduction" width="60" alt="Dimension reduction"></td>
<td align="right"><img src="img%2Fbatch_correction.svg" title="batch correction" width="60" alt="batch correction"></td>
</tr>
<tr class="even">
<td align="left">ASW batch</td>
<td align="center"><img src="img%2Fcell_type.svg" title="cell-type variable" alt="cell-type variable"></td>
<td align="center"></td>
<td align="center"><img src="img%2Fembedding.svg" title="Dimension reduction" width="60" alt="Dimension reduction"></td>
<td align="right"><img src="img%2Fbatch_correction.svg" title="batch correction" width="60" alt="batch correction"></td>
</tr>
<tr class="odd">
<td align="left">ASW</td>
<td align="center"><img src="img%2Fcell_type.svg" title="cell-type variable" alt="cell-type variable"></td>
<td align="center"></td>
<td align="center"><img src="img%2Fembedding.svg" title="Dimension reduction" width="60" alt="Dimension reduction"></td>
<td align="right"><img src="img%2Fbio_conservation.svg" title="bio-conservation" width="60" alt="bio-conservation"></td>
</tr>
<tr class="even">
<td align="left">ARI</td>
<td align="center"><img src="img%2Fcell_type.svg" title="cell-type variable" alt="cell-type variable"></td>
<td align="center"><img src="img%2Fcell_cluster.svg" title="clustering variable" alt="clustering variable"></td>
<td align="center"></td>
<td align="right"><img src="img%2Fbio_conservation.svg" title="bio-conservation" width="60" alt="bio-conservation"></td>
</tr>
<tr class="odd">
<td align="left">NMI</td>
<td align="center"><img src="img%2Fcell_type.svg" title="cell-type variable" alt="cell-type variable"></td>
<td align="center"><img src="img%2Fcell_cluster.svg" title="clustering variable" alt="clustering variable"></td>
<td align="center"></td>
<td align="right"><img src="img%2Fbio_conservation.svg" title="bio-conservation" width="60" alt="bio-conservation"></td>
</tr>
<tr class="even">
<td align="left">cLISI</td>
<td align="center"><img src="img%2Fcell_type.svg" title="cell-type variable" alt="cell-type variable"></td>
<td align="center"></td>
<td align="center">
<img src="img%2Fembedding.svg" title="Dimension reduction" width="60" alt="Dimension reduction"> or
<img src="img%2Fnetwork.svg" title="KNN graph" width="60" alt="KNN graph">
</td>
<td align="right"><img src="img%2Fbio_conservation.svg" title="bio-conservation" width="60" alt="bio-conservation"></td>
</tr>
<tr class="odd">
<td align="left">iLISI</td>
<td align="center"><img src="img%2Fcell_type.svg" title="cell-type variable" alt="cell-type variable"></td>
<td align="center"></td>
<td align="center">
<img src="img%2Fembedding.svg" title="Dimension reduction" width="60" alt="Dimension reduction"> or
<img src="img%2Fnetwork.svg" title="KNN graph" width="60" alt="KNN graph">
</td>
<td align="right"><img src="img%2Fbatch_correction.svg" title="batch correction" width="60" alt="batch correction"></td>
</tr>
<tr class="even">
<td align="left">kBET</td>
<td align="center"><img src="img%2Fcell_type.svg" title="cell-type variable" alt="cell-type variable"></td>
<td align="center"></td>
<td align="center">
<img src="img%2Fembedding.svg" title="Dimension reduction" width="60" alt="Dimension reduction"> or
<img src="img%2Fnetwork.svg" title="KNN graph" width="60" alt="KNN graph">
</td>
<td align="right"><img src="img%2Fbatch_correction.svg" title="batch correction" width="60" alt="batch correction"></td>
</tr>
<tr class="odd">
<td align="left">Graph connectivity</td>
<td align="center">
<img src="img%2Fcell_type.svg" title="cell-type variable" alt="cell-type variable">
(<code>per.component = TRUE</code>)</td>
<td align="center"></td>
<td align="center"><img src="img%2Fnetwork.svg" title="KNN graph" width="60" alt="KNN graph"></td>
<td align="right"><img src="img%2Fbatch_correction.svg" title="batch correction" width="60" alt="batch correction"></td>
</tr>
</tbody>
</table>
<p>Most scores are computed on an embedding <img src="img%2Fembedding.svg" title="DimReduc object" width="30" alt="DimReduc object">
(<code>Seurat::DimReduc</code> object) or a graph <img src="img%2Fnetwork.svg" title="Graph or Neighbor object" width="30" alt="Graph or Neighbor object"> (<code>Seurat::Neighbor</code> or
<code>Seurat::Graph</code> object). The exceptions are ARI and NMI,
which compare two categorical variables thus don’t need anything else
than a cell-type and a cluster assignment variables.d anything else than
a cell-type and a cluster assignment variables.</p>
<p>Most scores are based on a cell type label variable. This consists in
an estimate of each cell’s type obtained by analysing each batch
separately or by using an automatic cell annotation algorithm. This
estimate of cell types must be of sufficient quality to be considered
suitable as ground truth.</p>
</div>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/cbib/Seurat-Integrate" class="external-link">SeuratIntegrate</a></span><span class="op">)</span></span></code></pre></div>
<p>We will use the small dataset of 200 immune liver cells and around
6,500 genes included in SeuratIntegrate.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"liver_small"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span>  <span class="co"># genes x cells</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 6534  200</span></span></code></pre>
<p>It comprises 4 donors from 2 studies. Among the donors, 2 are healthy
and 2 are suffering from a hepatocellular carcinoma (HCC).</p>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["First_author"],"name":[1],"type":["chr"],"align":["left"]},{"label":["ID_sample"],"name":[2],"type":["chr"],"align":["left"]},{"label":["Tissue_type"],"name":[3],"type":["chr"],"align":["left"]},{"label":["n"],"name":[4],"type":["int"],"align":["right"]}],"data":[{"1":"Saleh","2":"B_13_TL","3":"Liver_HCC","4":"49"},{"1":"Saleh","2":"B_18_TL","3":"Liver_HCC","4":"30"},{"1":"Song","2":"N_01_TL","3":"Liver_HCC_tumoral","4":"43"},{"1":"Song","2":"N_02_TL","3":"Liver_HCC_tumoral","4":"78"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Importantly, the Seurat object’s metadata also embeds cell type
annotation variables.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_small</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">13</span>, <span class="fl">15</span><span class="op">:</span><span class="fl">17</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["predicted.celltype.l1"],"name":[1],"type":["chr"],"align":["left"]},{"label":["predicted.celltype.l2"],"name":[2],"type":["chr"],"align":["left"]},{"label":["manual_cell_type"],"name":[3],"type":["chr"],"align":["left"]},{"label":["manual_cell_type_short"],"name":[4],"type":["chr"],"align":["left"]}],"data":[{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_AAACGCTAGTGTTCCA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_AACAGGGAGTATGCAA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_AAGCCATCACCGTGCA-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"B_13_TL_ACCCAAATCGATCCCT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_ACGCACGTCGCCAGTG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_ACGGTTATCGGTAGAG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_ACGTACATCTCCGAAA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_AGATGAAGTGCTCGTG-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_AGCGTATAGTCGCCCA-1"},{"1":"T/NK","2":"CD8 T","3":"GD_T","4":"other_T","_rn_":"B_13_TL_ATCACTTAGCGTTGTT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_ATGATCGCACTTGTCC-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_ATGCATGCACCAGCTG-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_ATGGATCGTGATACCT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_CAACAACGTGCAACGA-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_CATCGGGTCGTGGGAA-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_CGAATTGAGTCTAGCT-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_CGGGTCAGTGTCATCA-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_CTAAGTGCAACGCATT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_CTCTGGTTCGCAGTTA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_CTTCCGACAAGGTTGG-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_GAGATGGTCAATCTCT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_GAGTGAGTCGGTGAAG-1"},{"1":"T/NK","2":"CD4 T","3":"GD_T","4":"other_T","_rn_":"B_13_TL_GATCAGTAGGACGGAG-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_GATGACTGTATTTCGG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GCATGATTCGCGTGCA-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GCCAGGTAGTGGTTAA-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GCCAGGTCACACCTTC-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GGAATCTTCGTTCTAT-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GGACGTCTCAGCGTCG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GGGACCTAGTAAGAGG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GGGTCACCAGGATGAC-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GGTCACGTCCCGAATA-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GGTCTGGCATCGTGCG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GTAATGCGTTTCGACA-1"},{"1":"Myeloid","2":"pDC","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_GTTCCGTTCGGTCACG-1"},{"1":"T/NK","2":"CD4 T","3":"GD_T","4":"other_T","_rn_":"B_13_TL_TACACCCTCTCTATAC-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_TATATCCTCCTACCGT-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_TCAAGTGAGTGATTCC-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_TCAGTGACACAACGAG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_TCATGTTCAATGGCAG-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"B_13_TL_TCCACGTAGTCCTGCG-1"},{"1":"T/NK","2":"CD4 T","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_TCGGGCAAGCCATCCG-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_TCGTAGATCATGTCAG-1"},{"1":"Myeloid","2":"DC2","3":"Macrophage_DC","4":"MNP","_rn_":"B_13_TL_TCTCACGGTTTACGAC-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_TGAATCGTCAACCTTT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_13_TL_TGTGAGTTCCAAGCTA-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_13_TL_TGTTGGACATCAGTGT-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"B_13_TL_TTCCTTCCAGTTGGTT-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"B_13_TL_TTGAACGAGTTGGCTT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_18_TL_AACACACGTCAACCTA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_18_TL_AATAGAGCAAGTTCGT-1"},{"1":"T/NK","2":"CD4 T","3":"GD_T","4":"other_T","_rn_":"B_18_TL_AATCGTGCACGCGTCA-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"B_18_TL_ACATGCAGTTTGTGGT-1"},{"1":"T/NK","2":"NK","3":"GD_T","4":"other_T","_rn_":"B_18_TL_ACTACGATCACTACGA-1"},{"1":"T/NK","2":"CD4 T","3":"GD_T","4":"other_T","_rn_":"B_18_TL_ACTTTGTAGTCACAGG-1"},{"1":"T/NK","2":"CD4 T","3":"NK","4":"NK","_rn_":"B_18_TL_AGATAGAAGACTCGAG-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_18_TL_AGCTACACATCGCTAA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_18_TL_ATAGGCTTCAGTCCGG-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_18_TL_ATCGATGAGACAGCGT-1"},{"1":"T/NK","2":"CD4 T","3":"GD_T","4":"other_T","_rn_":"B_18_TL_CCACGTTGTAAGGCCA-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"B_18_TL_CCACTTGTCGGTCTAA-1"},{"1":"Myeloid","2":"Monocyte","3":"Macrophage_DC","4":"MNP","_rn_":"B_18_TL_CGAGAAGCAGCAGAAC-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_18_TL_CTAACCCTCCTAGCCT-1"},{"1":"T/NK","2":"CD4 T","3":"GD_T","4":"other_T","_rn_":"B_18_TL_CTCCACACAAAGCAAT-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_18_TL_GACTCTCGTCTAATCG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_18_TL_GTTACAGTCAGTCACA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_18_TL_GTTGCTCCAGACTCTA-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"B_18_TL_GTTGTAGCACGTCATA-1"},{"1":"T/NK","2":"CD4 T","3":"GD_T","4":"other_T","_rn_":"B_18_TL_TAATCTCGTAAGACCG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_18_TL_TACAACGTCGTTTACT-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"B_18_TL_TACTTACGTGGATACG-1"},{"1":"T/NK","2":"NK_CD56bright","3":"GD_T","4":"other_T","_rn_":"B_18_TL_TATCTTGGTTTGCAGT-1"},{"1":"T/NK","2":"CD4 T","3":"GD_T","4":"other_T","_rn_":"B_18_TL_TCGCAGGAGGCTCTAT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_18_TL_TGCTTCGTCGTCGCTT-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"B_18_TL_TGTAACGGTAGAGCTG-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"B_18_TL_TGTCCTGCATTGACTG-1"},{"1":"T/NK","2":"CD4 T","3":"NK","4":"NK","_rn_":"B_18_TL_TTCATGTGTTGCCAAT-1"},{"1":"Myeloid","2":"Monocyte","3":"Macrophage_DC","4":"MNP","_rn_":"B_18_TL_TTCTTGATCGACCTAA-1"},{"1":"T/NK","2":"CD8 T","3":"GD_T","4":"other_T","_rn_":"B_18_TL_TTGACCCGTCGTTGGC-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_AATAGAGCATCCTATT-1"},{"1":"Myeloid","2":"pDC","3":"DC2","4":"MNP","_rn_":"N_01_TL_AATTCCTGTGGATCGA-1"},{"1":"T/NK","2":"CD8 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_ACCGTTCTCCGATCTC-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_ACGGAAGAGTCTCCTC-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_ACTTCCGAGCGACAGT-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_AGAACAAAGCATCTTG-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_AGACTCATCATATGGC-1"},{"1":"T/NK","2":"CD8 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_AGATGAAAGGGCTGAT-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_AGATGAAGTAACAAGT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_01_TL_AGGAAATGTTGTACGT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_01_TL_ATAGAGAAGTGGCGAT-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_ATGCCTCAGATGGTAT-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_ATGCGATTCGTCTAAG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"N_01_TL_ATTCACTAGGTTAAAC-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_CAGCACGCAACAGCTT-1"},{"1":"Myeloid","2":"DC1","3":"DC1","4":"MNP","_rn_":"N_01_TL_CAGTGCGTCCACGGAC-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_CCAATGACAGATGCGA-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_CTCATGCTCTCCGATC-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_01_TL_CTGGACGCAGACCTAT-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"N_01_TL_GAAACCTGTCTACGTA-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_GAACGTTTCCTCACGT-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_GAACTGTTCGCATGAT-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_GAAGAATCACGTACAT-1"},{"1":"T/NK","2":"NK","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_GAATCACAGATCGGTG-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_GAGCCTGGTTCTGACA-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_01_TL_GCATCGGTCGGAATTC-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_GCCAGTGAGCCGAACA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_01_TL_GGGATCCAGCACCGTC-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_GTCAGCGGTATCACCA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_01_TL_GTCTGTCCAAATTGCC-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_GTGCACGGTAACCCGC-1"},{"1":"T/NK","2":"CD8 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_GTGTAACTCTGTCCGT-1"},{"1":"T/NK","2":"CD8 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_TACAACGCAGCGTGAA-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_TCAGCAACAATTAGGA-1"},{"1":"T/NK","2":"CD8 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_TCCAGAAGTAGAGTTA-1"},{"1":"T/NK","2":"NK","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_TCTCACGGTCCCGCAA-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_01_TL_TGACTCCAGGCGTTGA-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"N_01_TL_TGATCAGCAAATCGTC-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_01_TL_TGCTTCGTCAAGCCCG-1"},{"1":"Myeloid","2":"DC2","3":"DC2","4":"MNP","_rn_":"N_01_TL_TGGAGGACATGGCCAC-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_TGGATCATCCTTCACG-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_01_TL_TTGGTTTTCAAACCCA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_01_TL_TTTCAGTGTGATACAA-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_AACACACTCTGCCCTA-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_AAGCATCCAAGGCTTT-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_02_TL_AATGAAGTCGATTGAC-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_ACACGCGAGTTAGTGA-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"N_02_TL_ACAGGGACAACATCGT-1"},{"1":"Myeloid","2":"DC2","3":"DC2","4":"MNP","_rn_":"N_02_TL_ACATCGAGTCGAAACG-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_ACGGTCGGTCGGTACC-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_ACGGTCGTCGGTCACG-1"},{"1":"Myeloid","2":"Monocyte","3":"Monocyte","4":"MNP","_rn_":"N_02_TL_ACTGTCCCACATCCCT-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_AGGAAATTCTCAACGA-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_AGGGAGTCAGGTCAGA-1"},{"1":"Myeloid","2":"Inflammatory Macrophage","3":"Monocyte","4":"MNP","_rn_":"N_02_TL_ATAGACCCACGGGTAA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_CAACGGCGTGATTCAC-1"},{"1":"Myeloid","2":"DC2","3":"DC2","4":"MNP","_rn_":"N_02_TL_CAAGGGACAGCGTGAA-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_CACGGGTCAGCGTAGA-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_CACTGTCGTGGACTAG-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_CAGAGCCAGATTAGCA-1"},{"1":"Myeloid","2":"DC2","3":"Monocyte","4":"MNP","_rn_":"N_02_TL_CAGATACCAGTGTGCC-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_CAGTTAGGTCTCGACG-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_CATCAAGCAATGAACA-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_CATCCACCAGACTCTA-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_02_TL_CATCCACGTGCATCTA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_CATCCCACAAGGTCTT-1"},{"1":"Myeloid","2":"DC2","3":"DC2","4":"MNP","_rn_":"N_02_TL_CATTCATTCAGCATTG-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_CCCAACTCATGGGCAA-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_CCCTCTCTCCCAGGCA-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_CCCTGATAGTGAGTTA-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_CCTAAGAAGATCGACG-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_02_TL_CCTGTTGTCTCGGTCT-1"},{"1":"Myeloid","2":"Hepatocyte","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_CGAGAAGAGAGGGTCT-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_CGAGAAGAGTAGACAT-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_02_TL_CGTCCATGTAGAGCTG-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_CTACGGGCAATAAGGT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_CTATAGGTCTGTTCAT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_CTCATCGTCAGGGTAG-1"},{"1":"T/NK","2":"NK","3":"GD_T","4":"other_T","_rn_":"N_02_TL_CTTCCTTGTCGGTGTC-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_GAAGGACAGGCCATAG-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_GAAGGGTGTCCTGTCT-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_02_TL_GACACGCTCGCTCTAC-1"},{"1":"T/NK","2":"CD4 T","3":"GD_T","4":"other_T","_rn_":"N_02_TL_GACATCAGTTTGGGTT-1"},{"1":"T/NK","2":"CD4 T","3":"NK","4":"NK","_rn_":"N_02_TL_GACCGTGTCTATTGTC-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_GACTGATGTGAAGCTG-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_GAGATGGAGGCTAAAT-1"},{"1":"Myeloid","2":"DC2","3":"DC2","4":"MNP","_rn_":"N_02_TL_GAGTTTGAGGGCTTCC-1"},{"1":"Myeloid","2":"DC2","3":"DC2","4":"MNP","_rn_":"N_02_TL_GCCCAGATCTTGATTC-1"},{"1":"Hepatocyte","2":"Hepatocyte","3":"Macrophage","4":"MNP","_rn_":"N_02_TL_GCGTTTCGTATCGAGG-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_GCTGCAGCACTCGATA-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_GCTGGGTCAACACGAG-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_GGGATCCGTGTCCATA-1"},{"1":"T/NK","2":"NK","3":"GD_T","4":"other_T","_rn_":"N_02_TL_GGGCTACTCCGTTGAA-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_GGGCTCACAAGTTCCA-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_GGTGATTGTAGCTTGT-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_GTAACACTCGGTATGT-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_GTAGCTAGTGAACGGT-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_GTCAAACAGCCTCAGC-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_GTCACGGAGCGGCTCT-1"},{"1":"Myeloid","2":"DC1","3":"Monocyte","4":"MNP","_rn_":"N_02_TL_GTCATTTTCCCTCTTT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_GTCGCGAGTGTGTGGA-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_GTTAGTGTCCACAGCG-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_TAATTCCAGTCCCGAC-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_TACTTACGTGTAGGAC-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_TAGACCAAGCGTTCAT-1"},{"1":"T/NK","2":"NK","3":"GD_T","4":"other_T","_rn_":"N_02_TL_TCCACGTGTTAGTCGT-1"},{"1":"T/NK","2":"NK","3":"NK","4":"NK","_rn_":"N_02_TL_TCGACGGCACCTGCAG-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_02_TL_TCTATCACACTACACA-1"},{"1":"T/NK","2":"NK","3":"GD_T","4":"other_T","_rn_":"N_02_TL_TCTTGCGCACTGTCCT-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_TGAGACTTCCGCTAGG-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_TGAGCGCCAGGAATAT-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_TGAGGAGCATTCTTCA-1"},{"1":"Myeloid","2":"Inflammatory Macrophage","3":"Monocyte","4":"MNP","_rn_":"N_02_TL_TGCGGGTAGTAGCTCT-1"},{"1":"T/NK","2":"NK","3":"GD_T","4":"other_T","_rn_":"N_02_TL_TGTGGCGGTGCCCGTA-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_TGTTTGTCACTGGCCA-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_02_TL_TTACGTTGTCGGTACC-1"},{"1":"T/NK","2":"NK_CD56bright","3":"NK","4":"NK","_rn_":"N_02_TL_TTATTGCCAGCTCTGG-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_TTGCCTGGTTACGATC-1"},{"1":"T/NK","2":"CD4 T","3":"T","4":"other_T","_rn_":"N_02_TL_TTGTTCAGTCGTTTCC-1"},{"1":"T/NK","2":"CD4 T","3":"CD4 T","4":"CD4 T","_rn_":"N_02_TL_TTTCAGTTCTCTAGGA-1"},{"1":"T/NK","2":"CD4 T","3":"CD8 T","4":"CD8 T","_rn_":"N_02_TL_TTTGGTTCAACGGGTA-1"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Note that albeit not compulsory, a <strong>high quality cell-type
annotation</strong> is very important to evaluate the performance of
integrations because many scoring metrics use them as ground truth.
Moreover, it helps in verifying whether technical effects (usually
called batch effects) overweight true biological differences.</p>
<p>Here, we use the last one and save it:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell.var</span> <span class="op">&lt;-</span> <span class="st">"manual_cell_type_short"</span></span></code></pre></div>
<p>We are going to test 3 integration methods that output different
types of objects. We will use ComBat (corrected counts), BBKNN
(corrected knn graph) and Harmony (corrected dimension reduction). BBKNN
is a Python method, so we need to have a conda environment to be able to
use it. Right now, we don’t have such an environment:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getCache.html">getCache</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="img%2Fconda_envs_unset.png" alt="Unset conda environments" width="100%" style="display: block; margin: auto;"></p>
<p>Hopefully, SeuratIntegrate facilitates the task with
<code><a href="../reference/UpdateEnvCache.html">UpdateEnvCache()</a></code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/UpdateEnvCache.html">UpdateEnvCache</a></span><span class="op">(</span><span class="st">"bbknn"</span><span class="op">)</span></span></code></pre></div>
<blockquote style="font-size: 1em;">
<em>Note</em>: Similarly, you can call <code><a href="../reference/UpdateEnvCache.html">UpdateEnvCache()</a></code>
with all other Python methods (“scvi”, “scanorama”, “trVAE”) to set up
their corresponding conda environments.
</blockquote>
<p>If the process is successful, the cache of conda environments should
have been updated:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getCache.html">getCache</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="img%2Fconda_envs_bbknn.png" alt="Conda environment set for BBKNN" width="100%" style="display: block; margin: auto;"></p>
<p>We also want to make sure we have a conda environment with a
umap-learn version compatible with Seurat:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/conda-tools.html" class="external-link">conda_create</a></span><span class="op">(</span><span class="st">'umap_0.5.4'</span>, packages <span class="op">=</span> <span class="st">'umap=0.5.4'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preprocessing-with-seurat">Preprocessing with Seurat<a class="anchor" aria-label="anchor" href="#preprocessing-with-seurat"></a>
</h2>
<p>To ensure that SeuratIntegrate works well, it is indispensable to
split the Seurat object. This process distributes cells into
single-batch layers such that each batch is normalized independently. We
account for potential technical variability between samples by
designating “ID_sample” as the batch variable.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">batch.var</span> <span class="op">&lt;-</span> <span class="st">"ID_sample"</span>  <span class="co"># save for later</span></span>
<span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">liver_small</span>, f <span class="op">=</span> <span class="va">liver_small</span><span class="op">$</span><span class="va">ID_sample</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span></span>
<span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">liver_small</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, k.param <span class="op">=</span> <span class="fl">15L</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">liver_small</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, n.neighbors <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">liver_small</span>, group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"First_author"</span>, <span class="st">"ID_sample"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SeuratIntegrate_files/figure-html/unnamed-chunk-13-1.png" alt="UMAP of unintegrated liver data with cells colored by first author and sample id (i.e. batch)" width="864"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">liver_small</span>, group.by <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span></span></code></pre></div>
<p><img src="SeuratIntegrate_files/figure-html/unnamed-chunk-14-1.png" alt="UMAP of unintegrated liver data with cells colored by cell type label" width="480"></p>
</div>
<div class="section level2">
<h2 id="integration-with-seuratintegrate">Integration with SeuratIntegrate<a class="anchor" aria-label="anchor" href="#integration-with-seuratintegrate"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_small</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="../reference/DoIntegrate.html">DoIntegrate</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">liver_small</span>,</span>
<span>              </span>
<span>            <span class="co"># integrations</span></span>
<span>              <span class="fu"><a href="../reference/CombatIntegration.html">CombatIntegration</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>              <span class="fu"><a href="../reference/bbknnIntegration.html">bbknnIntegration</a></span><span class="op">(</span>orig <span class="op">=</span> <span class="st">"pca"</span>, ndims.use <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,</span>
<span>              <span class="fu">SeuratIntegrate</span><span class="fu">::</span><span class="fu"><a href="../reference/HarmonyIntegration.html">HarmonyIntegration</a></span><span class="op">(</span>orig <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span>,</span>
<span>              </span>
<span>            <span class="co"># additional parameters</span></span>
<span>              use.hvg <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>              use.future <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Integration 1 in 3: integrating using 'CombatIntegration'</span></span>
<span><span class="co">## Integration 2 in 3: integrating using 'bbknnIntegration'</span></span>
<span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 200</span></span>
<span><span class="co">## Number of edges: 1784</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.6251</span></span>
<span><span class="co">## Number of communities: 5</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span>
<span><span class="co">## Integration 3 in 3: integrating using 'SeuratIntegrate::HarmonyIntegration'</span></span></code></pre>
<p>If we take a look at our Seurat object, we can note that it has been
enriched with many objects:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"\n#####\n\n"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu">Graphs</span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 22602 features across 200 samples within 4 assays </span></span>
<span><span class="co">## Active assay: SCT (6534 features, 3000 variable features)</span></span>
<span><span class="co">##  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">##  3 other assays present: RNA, combat.reconstructed, bbknn.ridge</span></span>
<span><span class="co">##  4 dimensional reductions calculated: pca, umap, pca.bbknn, harmony</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## #####</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [1] "SCT_nn"                              </span></span>
<span><span class="co">## [2] "SCT_snn"                             </span></span>
<span><span class="co">## [3] "bbknn_scale.data_connectivities"     </span></span>
<span><span class="co">## [4] "bbknn_scale.data_distances"          </span></span>
<span><span class="co">## [5] "bbknn_ridge.residuals_connectivities"</span></span>
<span><span class="co">## [6] "bbknn_ridge.residuals_distances"</span></span></code></pre>
<p>Important outputs are:</p>
<ul>
<li>New assay <strong>combat.reconstructed</strong>
</li>
<li>New reduction <strong>harmony</strong>
</li>
<li>New graphs <strong>bbknn_ridge.residuals_[…]</strong>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="post-processing">Post-processing<a class="anchor" aria-label="anchor" href="#post-processing"></a>
</h2>
<p>The type of output is important to consider, because scoring metrics
are not compatible with all output types. The simplest strategy is to
process each output separately in order to obtain at least a PCA out of
it, or even a knn graph (indispensable to compute clusters). Note that
several scores cannot be computed on knn graphs, hence knn graph outputs
(e.g. BBKNN) can only be evaluated by a reduced set of metrics.</p>
<p>Below is a summary for each output type (bracketed steps are not
always necessary):</p>
<ul>
<li>corrected counts: [<code><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData()</a></code>] -&gt;
<code><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA()</a></code> -&gt; [<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors()</a></code> -&gt;
<code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code>]</li>
<li>Dimension reduction: [<code><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA()</a></code>] -&gt;
[<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors()</a></code> -&gt;
<code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code>]</li>
<li>knn graph: [<code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code>]</li>
</ul>
<p>Here, we will use <code><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn()</a></code> between
<code><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors()</a></code> and <code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code>
because <code>return.neighbor = TRUE</code>. This is useful to keep the
distances between cells in the <code>KNN</code> graph. Although not
compulsory, this is used to stay in line with BBKNN’s output. To prevent
the community detection algorithm to output a high fraction of
singletons, we “<em>symmetrize</em>” the matrix which makes the graph
“undirected”.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># corrected counts outputs</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"combat.reconstructed"</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">[[</span><span class="st">"SCT"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span></span>
<span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">liver_small</span>, npcs <span class="op">=</span> <span class="fl">50L</span>, reduction.name <span class="op">=</span> <span class="st">"pca.combat"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">liver_small</span>, reduction <span class="op">=</span> <span class="st">"pca.combat"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,</span>
<span>                     return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>, graph.name <span class="op">=</span> <span class="st">"knn.combat"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">liver_small</span>, graph.name <span class="op">=</span> <span class="st">"knn.combat"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">liver_small</span>, graph.name <span class="op">=</span> <span class="st">"knn.combat_symmetric"</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"combat_{cell.var}_{metric}"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           optimisation.metric <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nmi"</span>, <span class="st">"ari"</span><span class="op">)</span><span class="op">)</span> <span class="co"># default, compute both</span></span>
<span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"SCT"</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># dimension reduction outputs</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">liver_small</span>, reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, k.param <span class="op">=</span> <span class="fl">20L</span>,</span>
<span>                     return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>, graph.name <span class="op">=</span> <span class="st">"knn.unintegrated"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">liver_small</span>, graph.name <span class="op">=</span> <span class="st">"knn.unintegrated"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">liver_small</span>, graph.name <span class="op">=</span> <span class="st">"knn.unintegrated_symmetric"</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"unintegrated_{cell.var}_{metric}"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">liver_small</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>,</span>
<span>                     return.neighbor <span class="op">=</span> <span class="cn">TRUE</span>, graph.name <span class="op">=</span> <span class="st">"knn.harmony"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">liver_small</span>, graph.name <span class="op">=</span> <span class="st">"knn.harmony"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">liver_small</span>, graph.name <span class="op">=</span> <span class="st">"knn.harmony_symmetric"</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           cluster.name <span class="op">=</span> <span class="st">"harmony_{cell.var}_{metric}"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># graph outputs</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SymmetrizeKnn.html">SymmetrizeKnn</a></span><span class="op">(</span><span class="va">liver_small</span>, graph.name <span class="op">=</span> <span class="st">"bbknn_ridge.residuals_distances"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindOptimalClusters.html">FindOptimalClusters</a></span><span class="op">(</span><span class="va">liver_small</span>, graph.name <span class="op">=</span> <span class="st">"bbknn_ridge.residuals_distances_symmetric"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>, cluster.name <span class="op">=</span> <span class="st">"bbknn_{cell.var}_{metric}"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scoring-1">Scoring<a class="anchor" aria-label="anchor" href="#scoring-1"></a>
</h2>
<p>Now that we have post-processed each integration’s output, we can
compute multiple scores to estimate the accuracy of the
integrations.</p>
<div class="section level3">
<h3 id="scores-not-requiring-a-cell-type-variable">Scores not requiring a cell type variable<a class="anchor" aria-label="anchor" href="#scores-not-requiring-a-cell-type-variable"></a>
</h3>
<p>Those scores are run on a dimension reduction (such as a PCA) and are
not compatible with graph outputs. They are cell-type free score thus do
not require a cell-type label variable.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-regressPC.html">AddScoreRegressPC</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>,</span>
<span>                                 batch.var <span class="op">=</span> <span class="va">batch.var</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-regressPC.html">AddScoreRegressPC</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"combat"</span>,</span>
<span>                                 batch.var <span class="op">=</span> <span class="va">batch.var</span>, reduction <span class="op">=</span> <span class="st">"pca.combat"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-regressPC.html">AddScoreRegressPC</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                                 batch.var <span class="op">=</span> <span class="va">batch.var</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-densityPC.html">AddScoreDensityPC</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>,</span>
<span>                                 batch.var <span class="op">=</span> <span class="va">batch.var</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-densityPC.html">AddScoreDensityPC</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"combat"</span>,</span>
<span>                                 batch.var <span class="op">=</span> <span class="va">batch.var</span>, reduction <span class="op">=</span> <span class="st">"pca.combat"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-regressPC.html">AddScoreRegressPC</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                                 batch.var <span class="op">=</span> <span class="va">batch.var</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CellCycleScoringPerBatch.html">CellCycleScoringPerBatch</a></span><span class="op">(</span><span class="va">liver_small</span>, batch.var <span class="op">=</span> <span class="va">batch.var</span>,</span>
<span>                                        s.features <span class="op">=</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">s.genes</span>,</span>
<span>                                        g2m.features <span class="op">=</span> <span class="va">cc.genes</span><span class="op">$</span><span class="va">g2m.genes</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-cc.html">AddScoreRegressPC.CellCycle</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>,</span>
<span>                                           batch.var <span class="op">=</span> <span class="va">batch.var</span>, what <span class="op">=</span> <span class="st">"pca"</span>,</span>
<span>                                           compute.cc <span class="op">=</span> <span class="cn">FALSE</span>, dims.use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-cc.html">AddScoreRegressPC.CellCycle</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"combat"</span>,</span>
<span>                                           batch.var <span class="op">=</span> <span class="va">batch.var</span>, what <span class="op">=</span> <span class="st">"pca.combat"</span>,</span>
<span>                                           compute.cc <span class="op">=</span> <span class="cn">FALSE</span>, dims.use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-cc.html">AddScoreRegressPC.CellCycle</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                                           batch.var <span class="op">=</span> <span class="va">batch.var</span>, what <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                                           compute.cc <span class="op">=</span> <span class="cn">FALSE</span>, dims.use <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="scores-based-on-silhouette-width">Scores based on silhouette width<a class="anchor" aria-label="anchor" href="#scores-based-on-silhouette-width"></a>
</h3>
<p>Those scores return an average silhouette width (ASW), either per
cell type label or per batch. They are run on a dimension reduction
(such as a PCA) and are not compatible with graph outputs. They are
based on a cell-type label variable. However,
<code><a href="../reference/score-asw.html">ScoreASWBatch()</a></code> can be run in a cell-type label independent
manner with <code>per.cell.var = FALSE</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-asw.html">AddScoreASW</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>, what <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-asw.html">AddScoreASW</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"combat"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>, what <span class="op">=</span> <span class="st">"pca.combat"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-asw.html">AddScoreASW</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>, what <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-asw.html">AddScoreASWBatch</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>,</span>
<span>                                batch.var <span class="op">=</span> <span class="va">batch.var</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                                what <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-asw.html">AddScoreASWBatch</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"combat"</span>,</span>
<span>                                batch.var <span class="op">=</span> <span class="va">batch.var</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                                what <span class="op">=</span> <span class="st">"pca.combat"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-asw.html">AddScoreASWBatch</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                                batch.var <span class="op">=</span> <span class="va">batch.var</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                                what <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="scores-working-on-a-graph">Scores working on a graph<a class="anchor" aria-label="anchor" href="#scores-working-on-a-graph"></a>
</h3>
<p>Those scores are graph outputs. They require a cell-type label
variable.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-connectivity.html">AddScoreConnectivity</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>,</span>
<span>                                    graph.name <span class="op">=</span> <span class="st">"knn.unintegrated_symmetric"</span>,</span>
<span>                                    cell.var <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-connectivity.html">AddScoreConnectivity</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"combat"</span>,</span>
<span>                                    graph.name <span class="op">=</span> <span class="st">"knn.combat_symmetric"</span>,</span>
<span>                                    cell.var <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-connectivity.html">AddScoreConnectivity</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                                    graph.name <span class="op">=</span> <span class="st">"knn.harmony_symmetric"</span>,</span>
<span>                                    cell.var <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-connectivity.html">AddScoreConnectivity</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"bbknn"</span>,</span>
<span>                                    graph.name <span class="op">=</span> <span class="st">"bbknn_ridge.residuals_distances_symmetric"</span>,</span>
<span>                                    cell.var <span class="op">=</span> <span class="va">cell.var</span><span class="op">)</span></span>
<span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-lisi.html">AddScoreLISI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>,</span>
<span>                            batch.var <span class="op">=</span> <span class="va">batch.var</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                            reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-lisi.html">AddScoreLISI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"combat"</span>,</span>
<span>                            batch.var <span class="op">=</span> <span class="va">batch.var</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                            reduction <span class="op">=</span> <span class="st">"pca.combat"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-lisi.html">AddScoreLISI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                            batch.var <span class="op">=</span> <span class="va">batch.var</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                            reduction <span class="op">=</span> <span class="st">"harmony"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-lisi.html">AddScoreLISI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"bbknn"</span>,</span>
<span>                            batch.var <span class="op">=</span> <span class="va">batch.var</span>, cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                            reduction <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                            graph.name <span class="op">=</span> <span class="st">"bbknn_ridge.residuals_distances_symmetric"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="scores-comparing-two-categorical-variables">Scores comparing two categorical variables<a class="anchor" aria-label="anchor" href="#scores-comparing-two-categorical-variables"></a>
</h3>
<p>Those scores compare two categorical variables. Thus, they don’t need
anything else than a cell-type and a cluster assignment variables
(computed during <a href="#post-processing">Post-processing</a> with
<code><a href="../reference/FindOptimalClusters.html">FindOptimalClusters()</a></code>).</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-ari.html">AddScoreARI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           clust.var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"unintegrated"</span>, <span class="va">cell.var</span>, <span class="st">"ari"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-ari.html">AddScoreARI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"combat"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           clust.var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"combat"</span>, <span class="va">cell.var</span>, <span class="st">"ari"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-ari.html">AddScoreARI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           clust.var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"harmony"</span>, <span class="va">cell.var</span>, <span class="st">"ari"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-ari.html">AddScoreARI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"bbknn"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           clust.var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"bbknn"</span>, <span class="va">cell.var</span>, <span class="st">"ari"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-nmi.html">AddScoreNMI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"unintegrated"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           clust.var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"unintegrated"</span>, <span class="va">cell.var</span>, <span class="st">"nmi"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-nmi.html">AddScoreNMI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"combat"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           clust.var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"combat"</span>, <span class="va">cell.var</span>, <span class="st">"nmi"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-nmi.html">AddScoreNMI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           clust.var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"harmony"</span>, <span class="va">cell.var</span>, <span class="st">"nmi"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/score-nmi.html">AddScoreNMI</a></span><span class="op">(</span><span class="va">liver_small</span>, integration <span class="op">=</span> <span class="st">"bbknn"</span>,</span>
<span>                           cell.var <span class="op">=</span> <span class="va">cell.var</span>,</span>
<span>                           clust.var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"bbknn"</span>, <span class="va">cell.var</span>, <span class="st">"nmi"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="scaling-and-comparison">Scaling and comparison<a class="anchor" aria-label="anchor" href="#scaling-and-comparison"></a>
</h3>
<p>Now that we have obtained several scores per integration, we can
compare their performances. First, let’s scale the scores between zero
and one and standardize their direction (the closer to one, always the
better). This step makes things easier to compare and improves
interpretability.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ScaleScores.html">ScaleScores</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span></span></code></pre></div>
<p>Now, let’s plot the integrations.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotScores.html">PlotScores</a></span><span class="op">(</span><span class="va">liver_small</span><span class="op">)</span></span></code></pre></div>
<p><img src="SeuratIntegrate_files/figure-html/unnamed-chunk-21-1.png" alt="Dot plot to compare performance of integrations based on scores" width="864"></p>
<p>We can observe that 3 overall scores have been computed, namely batch
correction, bio-conservation and overall (the average of the last two).
To balance each score’s contribution to the overall scores, a min-max
rescaling is applied on each score after scaling, stretching the
original score’s bounds to zero and one.</p>
<p>It can be disabled to preview the original scores:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotScores.html">PlotScores</a></span><span class="op">(</span><span class="va">liver_small</span>, rescale <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="SeuratIntegrate_files/figure-html/unnamed-chunk-22-1.png" alt="Dot plot of scores without min-max rescaling" width="864"></p>
<p>We can also hide scores or integrations. For instance, “PCA.density”
and “PCA.regression” are hardly informative. Let’s exclude them and see
how it affects overall scores:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotScores.html">PlotScores</a></span><span class="op">(</span><span class="va">liver_small</span>, rescale <span class="op">=</span> <span class="cn">FALSE</span>, exclude.score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PCA.density"</span>, <span class="st">"PCA.regression"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SeuratIntegrate_files/figure-html/unnamed-chunk-23-1.png" alt="Dot plot of scores without min-max rescaling, excluding non-informative scores" width="768"></p>
<p>Interestingly, we obtained the same ranking of integrations as in the
first figure.</p>
</div>
</div>
<div class="section level2">
<h2 id="projecting-integrations-on-umaps">Projecting integrations on UMAPs<a class="anchor" aria-label="anchor" href="#projecting-integrations-on-umaps"></a>
</h2>
<p>It is also possible to project the cells on a UMAP computed for each
integration’s output, as a complementary (visual) inspection of
integration performances.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">liver_small</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"pca.combat"</span>,</span>
<span>                       reduction.name <span class="op">=</span> <span class="st">"umap.combat"</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">liver_small</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"harmony"</span>,</span>
<span>                       reduction.name <span class="op">=</span> <span class="st">"umap.harmony"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://future.futureverse.org" class="external-link">future</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">multisession</span><span class="op">)</span></span>
<span><span class="va">liver_small</span> <span class="op"><a href="https://future.futureverse.org/reference/future.html" class="external-link">%&lt;-%</a></span> <span class="op">{</span></span>
<span>  <span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/use_python.html" class="external-link">use_condaenv</a></span><span class="op">(</span><span class="st">'umap_0.5.4'</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">liver_small</span>, graph <span class="op">=</span> <span class="st">"bbknn_ridge.residuals_connectivities"</span>,</span>
<span>          umap.method <span class="op">=</span> <span class="st">"umap-learn"</span>, n.epochs <span class="op">=</span> <span class="fl">200L</span>,</span>
<span>          reduction.name <span class="op">=</span> <span class="st">"umap.bbknn"</span><span class="op">)</span> <span class="op">}</span></span>
<span><span class="va">liver_small</span></span>
<span><span class="fu"><a href="https://future.futureverse.org/reference/plan.html" class="external-link">plan</a></span><span class="op">(</span><span class="va">sequential</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">liver_small</span>, group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">batch.var</span>, <span class="va">cell.var</span><span class="op">)</span>, reduction <span class="op">=</span> <span class="st">"umap"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="SeuratIntegrate_files/figure-html/unnamed-chunk-25-1.png" alt="UMAP of unintegrated data" width="864"><p class="caption">
UMAP of unintegrated data
</p>
</div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">liver_small</span>, group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">batch.var</span>, <span class="va">cell.var</span><span class="op">)</span>, reduction <span class="op">=</span> <span class="st">"umap.combat"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="SeuratIntegrate_files/figure-html/unnamed-chunk-26-1.png" alt="UMAP of ComBat integration" width="864"><p class="caption">
UMAP of ComBat integration
</p>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">liver_small</span>, group.by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">batch.var</span>, <span class="va">cell.var</span><span class="op">)</span>, reduction <span class="op">=</span> <span class="st">"umap.bbknn"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="SeuratIntegrate_files/figure-html/unnamed-chunk-27-1.png" alt="UMAP of BBKNN integration" width="864"><p class="caption">
UMAP of BBKNN integration
</p>
</div>
<details><summary><strong>Session info</strong>
</summary><pre><code><span>  <span class="co">## R version 4.4.3 (2025-02-28)</span></span>
<span>  <span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span>  <span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## Matrix products: default</span></span>
<span>  <span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span>  <span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## locale:</span></span>
<span>  <span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span>  <span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span>  <span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span>  <span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## time zone: UTC</span></span>
<span>  <span class="co">## tzcode source: system (glibc)</span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## attached base packages:</span></span>
<span>  <span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## other attached packages:</span></span>
<span>  <span class="co">## [1] future_1.34.0         dplyr_1.1.4           SeuratIntegrate_0.4.0</span></span>
<span>  <span class="co">## [4] Seurat_5.2.1          SeuratObject_5.0.2    sp_2.2-0             </span></span>
<span>  <span class="co">## </span></span>
<span>  <span class="co">## loaded via a namespace (and not attached):</span></span>
<span>  <span class="co">##   [1] RcppAnnoy_0.0.22            splines_4.4.3              </span></span>
<span>  <span class="co">##   [3] later_1.4.1                 batchelor_1.22.0           </span></span>
<span>  <span class="co">##   [5] tibble_3.2.1                polyclip_1.10-7            </span></span>
<span>  <span class="co">##   [7] XML_3.99-0.18               fastDummies_1.7.5          </span></span>
<span>  <span class="co">##   [9] lifecycle_1.0.4             edgeR_4.4.2                </span></span>
<span>  <span class="co">##  [11] globals_0.16.3              lattice_0.22-6             </span></span>
<span>  <span class="co">##  [13] MASS_7.3-64                 backports_1.5.0            </span></span>
<span>  <span class="co">##  [15] magrittr_2.0.3              limma_3.62.2               </span></span>
<span>  <span class="co">##  [17] plotly_4.10.4               sass_0.4.9                 </span></span>
<span>  <span class="co">##  [19] rmarkdown_2.29              jquerylib_0.1.4            </span></span>
<span>  <span class="co">##  [21] yaml_2.3.10                 httpuv_1.6.15              </span></span>
<span>  <span class="co">##  [23] glmGamPoi_1.18.0            sctransform_0.4.1          </span></span>
<span>  <span class="co">##  [25] spam_2.11-1                 spatstat.sparse_3.1-0      </span></span>
<span>  <span class="co">##  [27] reticulate_1.41.0.1         DBI_1.2.3                  </span></span>
<span>  <span class="co">##  [29] cowplot_1.1.3               pbapply_1.7-2              </span></span>
<span>  <span class="co">##  [31] RColorBrewer_1.1-3          ResidualMatrix_1.16.0      </span></span>
<span>  <span class="co">##  [33] abind_1.4-8                 zlibbioc_1.52.0            </span></span>
<span>  <span class="co">##  [35] Rtsne_0.17                  GenomicRanges_1.58.0       </span></span>
<span>  <span class="co">##  [37] purrr_1.0.4                 BiocGenerics_0.52.0        </span></span>
<span>  <span class="co">##  [39] tweenr_2.0.3                rappdirs_0.3.3             </span></span>
<span>  <span class="co">##  [41] sva_3.54.0                  GenomeInfoDbData_1.2.13    </span></span>
<span>  <span class="co">##  [43] IRanges_2.40.1              S4Vectors_0.44.0           </span></span>
<span>  <span class="co">##  [45] ggrepel_0.9.6               irlba_2.3.5.1              </span></span>
<span>  <span class="co">##  [47] listenv_0.9.1               spatstat.utils_3.1-3       </span></span>
<span>  <span class="co">##  [49] genefilter_1.88.0           goftest_1.2-3              </span></span>
<span>  <span class="co">##  [51] RSpectra_0.16-2             annotate_1.84.0            </span></span>
<span>  <span class="co">##  [53] spatstat.random_3.3-3       fitdistrplus_1.2-2         </span></span>
<span>  <span class="co">##  [55] parallelly_1.42.0           pkgdown_2.1.1              </span></span>
<span>  <span class="co">##  [57] DelayedMatrixStats_1.28.1   codetools_0.2-20           </span></span>
<span>  <span class="co">##  [59] DelayedArray_0.32.0         ggforce_0.4.2              </span></span>
<span>  <span class="co">##  [61] scuttle_1.16.0              tidyselect_1.2.1           </span></span>
<span>  <span class="co">##  [63] UCSC.utils_1.2.0            farver_2.1.2               </span></span>
<span>  <span class="co">##  [65] ScaledMatrix_1.14.0         matrixStats_1.5.0          </span></span>
<span>  <span class="co">##  [67] stats4_4.4.3                spatstat.explore_3.4-2     </span></span>
<span>  <span class="co">##  [69] jsonlite_1.9.1              BiocNeighbors_2.0.1        </span></span>
<span>  <span class="co">##  [71] progressr_0.15.1            ggridges_0.5.6             </span></span>
<span>  <span class="co">##  [73] survival_3.8-3              systemfonts_1.2.1          </span></span>
<span>  <span class="co">##  [75] tools_4.4.3                 ragg_1.3.3                 </span></span>
<span>  <span class="co">##  [77] ica_1.0-3                   Rcpp_1.0.14                </span></span>
<span>  <span class="co">##  [79] glue_1.8.0                  gridExtra_2.3              </span></span>
<span>  <span class="co">##  [81] SparseArray_1.6.2           mgcv_1.9-1                 </span></span>
<span>  <span class="co">##  [83] xfun_0.51                   MatrixGenerics_1.18.1      </span></span>
<span>  <span class="co">##  [85] GenomeInfoDb_1.42.3         withr_3.0.2                </span></span>
<span>  <span class="co">##  [87] fastmap_1.2.0               digest_0.6.37              </span></span>
<span>  <span class="co">##  [89] rsvd_1.0.5                  R6_2.6.1                   </span></span>
<span>  <span class="co">##  [91] mime_0.13                   textshaping_1.0.0          </span></span>
<span>  <span class="co">##  [93] colorspace_2.1-1            scattermore_1.2            </span></span>
<span>  <span class="co">##  [95] tensor_1.5                  RSQLite_2.3.9              </span></span>
<span>  <span class="co">##  [97] spatstat.data_3.1-6         RhpcBLASctl_0.23-42        </span></span>
<span>  <span class="co">##  [99] tidyr_1.3.1                 generics_0.1.3             </span></span>
<span>  <span class="co">## [101] data.table_1.17.0           httr_1.4.7                 </span></span>
<span>  <span class="co">## [103] htmlwidgets_1.6.4           S4Arrays_1.6.0             </span></span>
<span>  <span class="co">## [105] uwot_0.2.3                  pkgconfig_2.0.3            </span></span>
<span>  <span class="co">## [107] gtable_0.3.6                blob_1.2.4                 </span></span>
<span>  <span class="co">## [109] lmtest_0.9-40               SingleCellExperiment_1.28.1</span></span>
<span>  <span class="co">## [111] XVector_0.46.0              htmltools_0.5.8.1          </span></span>
<span>  <span class="co">## [113] dotCall64_1.2               scales_1.3.0               </span></span>
<span>  <span class="co">## [115] Biobase_2.66.0              png_0.1-8                  </span></span>
<span>  <span class="co">## [117] lisi_1.0                    harmony_1.2.3              </span></span>
<span>  <span class="co">## [119] spatstat.univar_3.1-2       knitr_1.50                 </span></span>
<span>  <span class="co">## [121] reshape2_1.4.4              nlme_3.1-167               </span></span>
<span>  <span class="co">## [123] cachem_1.1.0                zoo_1.8-13                 </span></span>
<span>  <span class="co">## [125] stringr_1.5.1               KernSmooth_2.23-26         </span></span>
<span>  <span class="co">## [127] parallel_4.4.3              miniUI_0.1.1.1             </span></span>
<span>  <span class="co">## [129] AnnotationDbi_1.68.0        desc_1.4.3                 </span></span>
<span>  <span class="co">## [131] pillar_1.10.1               grid_4.4.3                 </span></span>
<span>  <span class="co">## [133] vctrs_0.6.5                 RANN_2.6.2                 </span></span>
<span>  <span class="co">## [135] promises_1.3.2              BiocSingular_1.22.0        </span></span>
<span>  <span class="co">## [137] distances_0.1.11            beachmat_2.22.0            </span></span>
<span>  <span class="co">## [139] xtable_1.8-4                cluster_2.1.8              </span></span>
<span>  <span class="co">## [141] evaluate_1.0.3              locfit_1.5-9.12            </span></span>
<span>  <span class="co">## [143] cli_3.6.4                   compiler_4.4.3             </span></span>
<span>  <span class="co">## [145] rlang_1.1.5                 crayon_1.5.3               </span></span>
<span>  <span class="co">## [147] future.apply_1.11.3         labeling_0.4.3             </span></span>
<span>  <span class="co">## [149] plyr_1.8.9                  forcats_1.0.0              </span></span>
<span>  <span class="co">## [151] fs_1.6.5                    stringi_1.8.4              </span></span>
<span>  <span class="co">## [153] viridisLite_0.4.2           deldir_2.0-4               </span></span>
<span>  <span class="co">## [155] BiocParallel_1.40.0         Biostrings_2.74.1          </span></span>
<span>  <span class="co">## [157] munsell_0.5.1               lazyeval_0.2.2             </span></span>
<span>  <span class="co">## [159] spatstat.geom_3.3-6         Matrix_1.7-2               </span></span>
<span>  <span class="co">## [161] RcppHNSW_0.6.0              patchwork_1.3.0            </span></span>
<span>  <span class="co">## [163] bit64_4.6.0-1               sparseMatrixStats_1.18.0   </span></span>
<span>  <span class="co">## [165] ggplot2_3.5.1               statmod_1.5.0              </span></span>
<span>  <span class="co">## [167] KEGGREST_1.46.0             shiny_1.10.0               </span></span>
<span>  <span class="co">## [169] SummarizedExperiment_1.36.0 ROCR_1.0-11                </span></span>
<span>  <span class="co">## [171] memoise_2.0.1               igraph_2.1.4               </span></span>
<span>  <span class="co">## [173] broom_1.0.7                 bslib_0.9.0                </span></span>
<span>  <span class="co">## [175] bit_4.6.0</span></span></code></pre>
</details>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Florian Specque, Domitille Chalopin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
